{"message":"re-assignment of immutable variable `v` (Ast)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/liveness-assign-imm-local-in-op-eq.rs","byte_start":539,"byte_end":545,"line_start":14,"line_end":14,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    v += 1; //~ ERROR re-assignment of immutable variable","highlight_start":5,"highlight_end":11}],"label":"re-assignment of immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/liveness-assign-imm-local-in-op-eq.rs","byte_start":501,"byte_end":506,"line_start":13,"line_end":13,"column_start":5,"column_end":10,"is_primary":false,"text":[{"text":"    v = 2;  //~ NOTE first assignment","highlight_start":5,"highlight_end":10}],"label":"first assignment to `v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"re-assignment of immutable variable `v` (Mir)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/liveness-assign-imm-local-in-op-eq.rs","byte_start":539,"byte_end":545,"line_start":14,"line_end":14,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    v += 1; //~ ERROR re-assignment of immutable variable","highlight_start":5,"highlight_end":11}],"label":"re-assignment of immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/liveness-assign-imm-local-in-op-eq.rs","byte_start":501,"byte_end":506,"line_start":13,"line_end":13,"column_start":5,"column_end":10,"is_primary":false,"text":[{"text":"    v = 2;  //~ NOTE first assignment","highlight_start":5,"highlight_end":10}],"label":"first assignment to `v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
