{"message":"cannot infer an appropriate lifetime due to conflicting requirements","code":{"code":"E0495","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/regions-nested-fns.rs","byte_start":552,"byte_end":554,"line_start":15,"line_end":15,"column_start":18,"column_end":20,"is_primary":true,"text":[{"text":"    let mut ay = &y; //~ ERROR E0495","highlight_start":18,"highlight_end":20}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"first, the lifetime cannot outlive the anonymous lifetime #2 defined on the body at 17:54...","code":null,"level":"note","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/regions-nested-fns.rs","byte_start":626,"byte_end":686,"line_start":17,"line_end":21,"column_start":54,"column_end":6,"is_primary":true,"text":[{"text":"    ignore::<Box<for<'z> FnMut(&'z isize)>>(Box::new(|z| {","highlight_start":54,"highlight_end":59},{"text":"        ay = x;","highlight_start":1,"highlight_end":16},{"text":"        ay = &y;","highlight_start":1,"highlight_end":17},{"text":"        ay = z;","highlight_start":1,"highlight_end":16},{"text":"    }));","highlight_start":1,"highlight_end":6}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null},{"message":"...so that reference does not outlive borrowed content","code":null,"level":"note","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/regions-nested-fns.rs","byte_start":678,"byte_end":679,"line_start":20,"line_end":20,"column_start":14,"column_end":15,"is_primary":true,"text":[{"text":"        ay = z;","highlight_start":14,"highlight_end":15}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null},{"message":"but, the lifetime must be valid for the anonymous lifetime #2 defined on the body at 23:68...","code":null,"level":"note","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/regions-nested-fns.rs","byte_start":758,"byte_end":866,"line_start":23,"line_end":27,"column_start":68,"column_end":6,"is_primary":true,"text":[{"text":"    ignore::< Box<for<'z> FnMut(&'z isize) -> &'z isize>>(Box::new(|z| {","highlight_start":68,"highlight_end":73},{"text":"        if false { return x; } //~ ERROR E0312","highlight_start":1,"highlight_end":47},{"text":"        if false { return ay; }","highlight_start":1,"highlight_end":32},{"text":"        return z;","highlight_start":1,"highlight_end":18},{"text":"    }));","highlight_start":1,"highlight_end":6}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null},{"message":"...so that types are compatible (expected &isize, found &isize)","code":null,"level":"note","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/regions-nested-fns.rs","byte_start":762,"byte_end":866,"line_start":23,"line_end":27,"column_start":72,"column_end":6,"is_primary":true,"text":[{"text":"    ignore::< Box<for<'z> FnMut(&'z isize) -> &'z isize>>(Box::new(|z| {","highlight_start":72,"highlight_end":73},{"text":"        if false { return x; } //~ ERROR E0312","highlight_start":1,"highlight_end":47},{"text":"        if false { return ay; }","highlight_start":1,"highlight_end":32},{"text":"        return z;","highlight_start":1,"highlight_end":18},{"text":"    }));","highlight_start":1,"highlight_end":6}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}],"rendered":null}
{"message":"lifetime of reference outlives lifetime of borrowed content...","code":{"code":"E0312","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/regions-nested-fns.rs","byte_start":790,"byte_end":791,"line_start":24,"line_end":24,"column_start":27,"column_end":28,"is_primary":true,"text":[{"text":"        if false { return x; } //~ ERROR E0312","highlight_start":27,"highlight_end":28}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"...the reference is valid for the anonymous lifetime #2 defined on the body at 23:68...","code":null,"level":"note","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/regions-nested-fns.rs","byte_start":758,"byte_end":866,"line_start":23,"line_end":27,"column_start":68,"column_end":6,"is_primary":true,"text":[{"text":"    ignore::< Box<for<'z> FnMut(&'z isize) -> &'z isize>>(Box::new(|z| {","highlight_start":68,"highlight_end":73},{"text":"        if false { return x; } //~ ERROR E0312","highlight_start":1,"highlight_end":47},{"text":"        if false { return ay; }","highlight_start":1,"highlight_end":32},{"text":"        return z;","highlight_start":1,"highlight_end":18},{"text":"    }));","highlight_start":1,"highlight_end":6}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null},{"message":"...but the borrowed content is only valid for the lifetime 'x as defined on the function body at 13:1","code":null,"level":"note","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/regions-nested-fns.rs","byte_start":490,"byte_end":871,"line_start":13,"line_end":28,"column_start":1,"column_end":2,"is_primary":true,"text":[{"text":"fn nested<'x>(x: &'x isize) {","highlight_start":1,"highlight_end":30},{"text":"    let y = 3;","highlight_start":1,"highlight_end":15},{"text":"    let mut ay = &y; //~ ERROR E0495","highlight_start":1,"highlight_end":37},{"text":"","highlight_start":1,"highlight_end":1},{"text":"    ignore::<Box<for<'z> FnMut(&'z isize)>>(Box::new(|z| {","highlight_start":1,"highlight_end":59},{"text":"        ay = x;","highlight_start":1,"highlight_end":16},{"text":"        ay = &y;","highlight_start":1,"highlight_end":17},{"text":"        ay = z;","highlight_start":1,"highlight_end":16},{"text":"    }));","highlight_start":1,"highlight_end":9},{"text":"","highlight_start":1,"highlight_end":1},{"text":"    ignore::< Box<for<'z> FnMut(&'z isize) -> &'z isize>>(Box::new(|z| {","highlight_start":1,"highlight_end":73},{"text":"        if false { return x; } //~ ERROR E0312","highlight_start":1,"highlight_end":47},{"text":"        if false { return ay; }","highlight_start":1,"highlight_end":32},{"text":"        return z;","highlight_start":1,"highlight_end":18},{"text":"    }));","highlight_start":1,"highlight_end":9},{"text":"}","highlight_start":1,"highlight_end":2}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
