{"message":"cannot assign to immutable item `fancy_ref.num`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/E0389.rs","byte_start":592,"byte_end":609,"line_start":18,"line_end":18,"column_start":5,"column_end":22,"is_primary":true,"text":[{"text":"    fancy_ref.num = 6; //~ ERROR E0389","highlight_start":5,"highlight_end":22}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*fancy_ref`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `fancy_ref.num`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/E0389.rs:18:5\n   |\n18 |     fancy_ref.num = 6; //~ ERROR E0389\n   |     ^^^^^^^^^^^^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `*fancy_ref`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
