{"message":"cannot borrow `i` as mutable more than once at a time","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/E0499.rs","byte_start":514,"byte_end":520,"line_start":13,"line_end":13,"column_start":17,"column_end":23,"is_primary":false,"text":[{"text":"    let mut x = &mut i;","highlight_start":17,"highlight_end":23}],"label":"first mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/E0499.rs","byte_start":538,"byte_end":544,"line_start":14,"line_end":14,"column_start":17,"column_end":23,"is_primary":true,"text":[{"text":"    let mut a = &mut i; //~ ERROR E0499","highlight_start":17,"highlight_end":23}],"label":"second mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/E0499.rs","byte_start":562,"byte_end":563,"line_start":15,"line_end":15,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"first borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0499]: cannot borrow `i` as mutable more than once at a time\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/E0499.rs:14:17\n   |\n13 |     let mut x = &mut i;\n   |                 ------ first mutable borrow occurs here\n14 |     let mut a = &mut i; //~ ERROR E0499\n   |                 ^^^^^^ second mutable borrow occurs here\n15 | }\n   | - first borrow ends here\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
