{"message":"cannot move out of `fancy_num` because it is borrowed","code":{"code":"E0505","explanation":"\nA value was moved out while it was still borrowed.\n\nErroneous code example:\n\n```compile_fail,E0505\nstruct Value {}\n\nfn eat(val: Value) {}\n\nfn main() {\n    let x = Value{};\n    {\n        let _ref_to_val: &Value = &x;\n        eat(x);\n    }\n}\n```\n\nHere, the function `eat` takes the ownership of `x`. However,\n`x` cannot be moved because it was borrowed to `_ref_to_val`.\nTo fix that you can do few different things:\n\n* Try to avoid moving the variable.\n* Release borrow before move.\n* Implement the `Copy` trait on the type.\n\nExamples:\n\n```\nstruct Value {}\n\nfn eat(val: &Value) {}\n\nfn main() {\n    let x = Value{};\n    {\n        let _ref_to_val: &Value = &x;\n        eat(&x); // pass by reference, if it's possible\n    }\n}\n```\n\nOr:\n\n```\nstruct Value {}\n\nfn eat(val: Value) {}\n\nfn main() {\n    let x = Value{};\n    {\n        let _ref_to_val: &Value = &x;\n    }\n    eat(x); // release borrow and then move it.\n}\n```\n\nOr:\n\n```\n#[derive(Clone, Copy)] // implement Copy trait\nstruct Value {}\n\nfn eat(val: Value) {}\n\nfn main() {\n    let x = Value{};\n    {\n        let _ref_to_val: &Value = &x;\n        eat(x); // it will be copied here.\n    }\n}\n```\n\nYou can find more information about borrowing in the rust-book:\nhttp://doc.rust-lang.org/stable/book/references-and-borrowing.html\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/E0504.rs","byte_start":574,"byte_end":584,"line_start":17,"line_end":17,"column_start":21,"column_end":31,"is_primary":false,"text":[{"text":"    let fancy_ref = &fancy_num;","highlight_start":21,"highlight_end":31}],"label":"borrow of `fancy_num` occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/E0504.rs","byte_start":599,"byte_end":685,"line_start":19,"line_end":21,"column_start":13,"column_end":6,"is_primary":true,"text":[{"text":"    let x = move || {","highlight_start":13,"highlight_end":22},{"text":"        println!(\"child function: {}\", fancy_num.num); //~ ERROR E0504","highlight_start":1,"highlight_end":71},{"text":"    };","highlight_start":1,"highlight_end":6}],"label":"move out of `fancy_num` occurs here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0505]: cannot move out of `fancy_num` because it is borrowed\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/E0504.rs:19:13\n   |\n17 |       let fancy_ref = &fancy_num;\n   |                       ---------- borrow of `fancy_num` occurs here\n18 | \n19 |       let x = move || {\n   |  _____________^\n20 | |         println!(\"child function: {}\", fancy_num.num); //~ ERROR E0504\n21 | |     };\n   | |_____^ move out of `fancy_num` occurs here\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
