{"message":"cannot assign to immutable `Box` content `*y` (Ast)","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":494,"byte_end":500,"line_start":13,"line_end":13,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    *y = 5; //~ ERROR cannot assign","highlight_start":5,"highlight_end":11}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":473,"byte_end":474,"line_start":12,"line_end":12,"column_start":6,"column_end":7,"is_primary":false,"text":[{"text":"fn f(y: Box<isize>) {","highlight_start":6,"highlight_end":7}],"label":"consider changing this to `mut y`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign to immutable `Box` content `*q` (Ast)","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":572,"byte_end":578,"line_start":17,"line_end":17,"column_start":35,"column_end":41,"is_primary":true,"text":[{"text":"    let _frob = |q: Box<isize>| { *q = 2; }; //~ ERROR cannot assign","highlight_start":35,"highlight_end":41}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":555,"byte_end":556,"line_start":17,"line_end":17,"column_start":18,"column_end":19,"is_primary":false,"text":[{"text":"    let _frob = |q: Box<isize>| { *q = 2; }; //~ ERROR cannot assign","highlight_start":18,"highlight_end":19}],"label":"consider changing this to `mut q`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign twice to immutable variable `(*y)` (Mir)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":494,"byte_end":500,"line_start":13,"line_end":13,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    *y = 5; //~ ERROR cannot assign","highlight_start":5,"highlight_end":11}],"label":"cannot assign twice to immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":494,"byte_end":500,"line_start":13,"line_end":13,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    *y = 5; //~ ERROR cannot assign","highlight_start":5,"highlight_end":11}],"label":"first assignment to `(*y)`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign twice to immutable variable `(*q)` (Mir)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":572,"byte_end":578,"line_start":17,"line_end":17,"column_start":35,"column_end":41,"is_primary":true,"text":[{"text":"    let _frob = |q: Box<isize>| { *q = 2; }; //~ ERROR cannot assign","highlight_start":35,"highlight_end":41}],"label":"cannot assign twice to immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":572,"byte_end":578,"line_start":17,"line_end":17,"column_start":35,"column_end":41,"is_primary":true,"text":[{"text":"    let _frob = |q: Box<isize>| { *q = 2; }; //~ ERROR cannot assign","highlight_start":35,"highlight_end":41}],"label":"first assignment to `(*q)`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
