{"message":"cannot assign to immutable item `*y`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":494,"byte_end":500,"line_start":13,"line_end":13,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    *y = 5; //~ ERROR cannot assign","highlight_start":5,"highlight_end":11}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `y`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `*y`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs:13:5\n   |\n13 |     *y = 5; //~ ERROR cannot assign\n   |     ^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `y`\n\n"}
{"message":"cannot assign to immutable item `*q`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs","byte_start":572,"byte_end":578,"line_start":17,"line_end":17,"column_start":35,"column_end":41,"is_primary":true,"text":[{"text":"    let _frob = |q: Box<isize>| { *q = 2; }; //~ ERROR cannot assign","highlight_start":35,"highlight_end":41}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `q`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `*q`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/immut-function-arguments.rs:17:35\n   |\n17 |     let _frob = |q: Box<isize>| { *q = 2; }; //~ ERROR cannot assign\n   |                                   ^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `q`\n\n"}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
