{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-21600.rs","byte_start":754,"byte_end":768,"line_start":24,"line_end":24,"column_start":17,"column_end":31,"is_primary":true,"text":[{"text":"        call_it(|| x.gen_mut()); //~ ERROR cannot borrow data mutably in a captured outer","highlight_start":17,"highlight_end":31}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/issue-21600.rs:24:17\n   |\n24 |         call_it(|| x.gen_mut()); //~ ERROR cannot borrow data mutably in a captured outer\n   |                 ^^^^^^^^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-21600.rs","byte_start":757,"byte_end":758,"line_start":24,"line_end":24,"column_start":20,"column_end":21,"is_primary":true,"text":[{"text":"        call_it(|| x.gen_mut()); //~ ERROR cannot borrow data mutably in a captured outer","highlight_start":20,"highlight_end":21}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/issue-21600.rs:24:20\n   |\n24 |         call_it(|| x.gen_mut()); //~ ERROR cannot borrow data mutably in a captured outer\n   |                    ^ cannot borrow as mutable\n\n"}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
