{"message":"borrowed value does not live long enough","code":{"code":"E0597","explanation":"\nThis error occurs because a borrow was made inside a variable which has a\ngreater lifetime than the borrowed one.\n\nExample of erroneous code:\n\n```compile_fail,E0597\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet mut x = Foo { x: None };\nlet y = 0;\nx.x = Some(&y); // error: `y` does not live long enough\n```\n\nIn here, `x` is created before `y` and therefore has a greater lifetime. Always\nkeep in mind that values in a scope are dropped in the opposite order they are\ncreated. So to fix the previous example, just make the `y` lifetime greater than\nthe `x`'s one:\n\n```\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet y = 0;\nlet mut x = Foo { x: None };\nx.x = Some(&y);\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-11493.rs","byte_start":610,"byte_end":615,"line_start":17,"line_end":17,"column_start":35,"column_end":40,"is_primary":false,"text":[{"text":"    let y = x.as_ref().unwrap_or(&id(5)); //~ ERROR: borrowed value does not live long enough","highlight_start":35,"highlight_end":40}],"label":"temporary value created here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-11493.rs","byte_start":617,"byte_end":617,"line_start":17,"line_end":17,"column_start":42,"column_end":42,"is_primary":true,"text":[{"text":"    let y = x.as_ref().unwrap_or(&id(5)); //~ ERROR: borrowed value does not live long enough","highlight_start":42,"highlight_end":42}],"label":"temporary value dropped here while still borrowed","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-11493.rs","byte_start":670,"byte_end":671,"line_start":18,"line_end":18,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"temporary value needs to live until here","suggested_replacement":null,"expansion":null}],"children":[{"message":"consider using a `let` binding to increase its lifetime","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0597]: borrowed value does not live long enough\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/issue-11493.rs:17:42\n   |\n17 |     let y = x.as_ref().unwrap_or(&id(5)); //~ ERROR: borrowed value does not live long enough\n   |                                   -----  ^ temporary value dropped here while still borrowed\n   |                                   |\n   |                                   temporary value created here\n18 | }\n   | - temporary value needs to live until here\n   |\n   = note: consider using a `let` binding to increase its lifetime\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
