{"message":"cannot assign twice to immutable variable `v` (Ast)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/liveness-assign-imm-local-in-loop.rs","byte_start":516,"byte_end":521,"line_start":14,"line_end":14,"column_start":9,"column_end":14,"is_primary":true,"text":[{"text":"        v = 1; //~ ERROR cannot assign twice to immutable variable","highlight_start":9,"highlight_end":14}],"label":"cannot assign twice to immutable variable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign twice to immutable variable `v` (Mir)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/liveness-assign-imm-local-in-loop.rs","byte_start":516,"byte_end":521,"line_start":14,"line_end":14,"column_start":9,"column_end":14,"is_primary":true,"text":[{"text":"        v = 1; //~ ERROR cannot assign twice to immutable variable","highlight_start":9,"highlight_end":14}],"label":"cannot assign twice to immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/liveness-assign-imm-local-in-loop.rs","byte_start":516,"byte_end":521,"line_start":14,"line_end":14,"column_start":9,"column_end":14,"is_primary":true,"text":[{"text":"        v = 1; //~ ERROR cannot assign twice to immutable variable","highlight_start":9,"highlight_end":14}],"label":"first assignment to `v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
