{"message":"use of possibly uninitialized variable: `a`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/move-into-dead-array-1.rs","byte_start":673,"byte_end":677,"line_start":24,"line_end":24,"column_start":5,"column_end":9,"is_primary":true,"text":[{"text":"    a[i] = d();        //~ ERROR use of possibly uninitialized variable: `a`","highlight_start":5,"highlight_end":9}],"label":"use of possibly uninitialized `a`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `a`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/move-into-dead-array-1.rs:24:5\n   |\n24 |     a[i] = d();        //~ ERROR use of possibly uninitialized variable: `a`\n   |     ^^^^ use of possibly uninitialized `a`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
