{"message":"use of possibly uninitialized variable: `a` (Ast)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/move-into-dead-array-1.rs","byte_start":673,"byte_end":683,"line_start":24,"line_end":24,"column_start":5,"column_end":15,"is_primary":true,"text":[{"text":"    a[i] = d();        //~ ERROR use of possibly uninitialized variable: `a`","highlight_start":5,"highlight_end":15}],"label":"use of possibly uninitialized `a`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"use of possibly uninitialized variable: `a` (Mir)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/move-into-dead-array-1.rs","byte_start":673,"byte_end":677,"line_start":24,"line_end":24,"column_start":5,"column_end":9,"is_primary":true,"text":[{"text":"    a[i] = d();        //~ ERROR use of possibly uninitialized variable: `a`","highlight_start":5,"highlight_end":9}],"label":"use of possibly uninitialized `a`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
