{"message":"cannot assign to immutable item `self.how_hungry`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/mutable-class-fields-2.rs","byte_start":560,"byte_end":580,"line_start":19,"line_end":19,"column_start":5,"column_end":25,"is_primary":true,"text":[{"text":"    self.how_hungry -= 5; //~ ERROR cannot assign","highlight_start":5,"highlight_end":25}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*self`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `self.how_hungry`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/mutable-class-fields-2.rs:19:5\n   |\n19 |     self.how_hungry -= 5; //~ ERROR cannot assign\n   |     ^^^^^^^^^^^^^^^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `*self`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
