{"message":"cannot assign to immutable field `nyan.how_hungry` (Ast)","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/mutable-class-fields.rs","byte_start":677,"byte_end":696,"line_start":25,"line_end":25,"column_start":3,"column_end":22,"is_primary":true,"text":[{"text":"  nyan.how_hungry = 0; //~ ERROR cannot assign","highlight_start":3,"highlight_end":22}],"label":"cannot mutably borrow immutable field","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/mutable-class-fields.rs","byte_start":649,"byte_end":653,"line_start":24,"line_end":24,"column_start":7,"column_end":11,"is_primary":false,"text":[{"text":"  let nyan : cat = cat(52, 99);","highlight_start":7,"highlight_end":11}],"label":"consider changing this to `mut nyan`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign twice to immutable variable `nyan.how_hungry` (Mir)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/mutable-class-fields.rs","byte_start":677,"byte_end":696,"line_start":25,"line_end":25,"column_start":3,"column_end":22,"is_primary":true,"text":[{"text":"  nyan.how_hungry = 0; //~ ERROR cannot assign","highlight_start":3,"highlight_end":22}],"label":"cannot assign twice to immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/mutable-class-fields.rs","byte_start":677,"byte_end":696,"line_start":25,"line_end":25,"column_start":3,"column_end":22,"is_primary":true,"text":[{"text":"  nyan.how_hungry = 0; //~ ERROR cannot assign","highlight_start":3,"highlight_end":22}],"label":"first assignment to `nyan.how_hungry`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
