{"message":"cannot assign to immutable item `nyan.how_hungry`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/mutable-class-fields.rs","byte_start":677,"byte_end":696,"line_start":25,"line_end":25,"column_start":3,"column_end":22,"is_primary":true,"text":[{"text":"  nyan.how_hungry = 0; //~ ERROR cannot assign","highlight_start":3,"highlight_end":22}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `nyan`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `nyan.how_hungry`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/mutable-class-fields.rs:25:3\n   |\n25 |   nyan.how_hungry = 0; //~ ERROR cannot assign\n   |   ^^^^^^^^^^^^^^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `nyan`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
