{"message":"borrowed value does not live long enough","code":{"code":"E0597","explanation":"\nThis error occurs because a borrow was made inside a variable which has a\ngreater lifetime than the borrowed one.\n\nExample of erroneous code:\n\n```compile_fail,E0597\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet mut x = Foo { x: None };\nlet y = 0;\nx.x = Some(&y); // error: `y` does not live long enough\n```\n\nIn here, `x` is created before `y` and therefore has a greater lifetime. Always\nkeep in mind that values in a scope are dropped in the opposite order they are\ncreated. So to fix the previous example, just make the `y` lifetime greater than\nthe `x`'s one:\n\n```\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet y = 0;\nlet mut x = Foo { x: None };\nx.x = Some(&y);\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":926,"byte_end":943,"line_start":33,"line_end":33,"column_start":17,"column_end":34,"is_primary":false,"text":[{"text":"        return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough","highlight_start":17,"highlight_end":34}],"label":"temporary value created here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":944,"byte_end":944,"line_start":33,"line_end":33,"column_start":35,"column_end":35,"is_primary":true,"text":[{"text":"        return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough","highlight_start":35,"highlight_end":35}],"label":"temporary value dropped here while still borrowed","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":1200,"byte_end":1201,"line_start":41,"line_end":41,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"temporary value needs to live until here","suggested_replacement":null,"expansion":null}],"children":[{"message":"consider using a `let` binding to increase its lifetime","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0597]: borrowed value does not live long enough\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs:33:35\n   |\n33 |         return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough\n   |                 ----------------- ^ temporary value dropped here while still borrowed\n   |                 |\n   |                 temporary value created here\n...\n41 | }\n   | - temporary value needs to live until here\n   |\n   = note: consider using a `let` binding to increase its lifetime\n\n"}
{"message":"borrowed value does not live long enough","code":{"code":"E0597","explanation":"\nThis error occurs because a borrow was made inside a variable which has a\ngreater lifetime than the borrowed one.\n\nExample of erroneous code:\n\n```compile_fail,E0597\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet mut x = Foo { x: None };\nlet y = 0;\nx.x = Some(&y); // error: `y` does not live long enough\n```\n\nIn here, `x` is created before `y` and therefore has a greater lifetime. Always\nkeep in mind that values in a scope are dropped in the opposite order they are\ncreated. So to fix the previous example, just make the `y` lifetime greater than\nthe `x`'s one:\n\n```\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet y = 0;\nlet mut x = Foo { x: None };\nx.x = Some(&y);\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":1114,"byte_end":1132,"line_start":38,"line_end":38,"column_start":17,"column_end":35,"is_primary":false,"text":[{"text":"        return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough","highlight_start":17,"highlight_end":35}],"label":"temporary value created here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":1133,"byte_end":1133,"line_start":38,"line_end":38,"column_start":36,"column_end":36,"is_primary":true,"text":[{"text":"        return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough","highlight_start":36,"highlight_end":36}],"label":"temporary value dropped here while still borrowed","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":1200,"byte_end":1201,"line_start":41,"line_end":41,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"temporary value needs to live until here","suggested_replacement":null,"expansion":null}],"children":[{"message":"consider using a `let` binding to increase its lifetime","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0597]: borrowed value does not live long enough\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs:38:36\n   |\n38 |         return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough\n   |                 ------------------ ^ temporary value dropped here while still borrowed\n   |                 |\n   |                 temporary value created here\n...\n41 | }\n   | - temporary value needs to live until here\n   |\n   = note: consider using a `let` binding to increase its lifetime\n\n"}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
