{"message":"borrowed value does not live long enough (Ast)","code":{"code":"E0597","explanation":"\nThis error occurs because a borrow was made inside a variable which has a\ngreater lifetime than the borrowed one.\n\nExample of erroneous code:\n\n```compile_fail,E0597\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet mut x = Foo { x: None };\nlet y = 0;\nx.x = Some(&y); // error: `y` does not live long enough\n```\n\nIn here, `x` is created before `y` and therefore has a greater lifetime. Always\nkeep in mind that values in a scope are dropped in the opposite order they are\ncreated. So to fix the previous example, just make the `y` lifetime greater than\nthe `x`'s one:\n\n```\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet y = 0;\nlet mut x = Foo { x: None };\nx.x = Some(&y);\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":926,"byte_end":943,"line_start":33,"line_end":33,"column_start":17,"column_end":34,"is_primary":true,"text":[{"text":"        return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough","highlight_start":17,"highlight_end":34}],"label":"does not live long enough","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":943,"byte_end":944,"line_start":33,"line_end":33,"column_start":34,"column_end":35,"is_primary":false,"text":[{"text":"        return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough","highlight_start":34,"highlight_end":35}],"label":"temporary value only lives until here","suggested_replacement":null,"expansion":null}],"children":[{"message":"borrowed value must be valid for the lifetime 'a as defined on the function body at 30:1...","code":null,"level":"note","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":782,"byte_end":1201,"line_start":30,"line_end":41,"column_start":1,"column_end":2,"is_primary":true,"text":[{"text":"fn map_nums<'a,'b, F>(x: &ast, f: &mut F) -> &'a ast<'b> where F: FnMut(usize) -> usize {","highlight_start":1,"highlight_end":90},{"text":"    match *x {","highlight_start":1,"highlight_end":15},{"text":"      ast::num(x) => {","highlight_start":1,"highlight_end":23},{"text":"        return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough","highlight_start":1,"highlight_end":86},{"text":"      }","highlight_start":1,"highlight_end":8},{"text":"      ast::add(x, y) => {","highlight_start":1,"highlight_end":26},{"text":"        let m_x = map_nums(x, f);","highlight_start":1,"highlight_end":34},{"text":"        let m_y = map_nums(y, f);","highlight_start":1,"highlight_end":34},{"text":"        return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough","highlight_start":1,"highlight_end":88},{"text":"      }","highlight_start":1,"highlight_end":8},{"text":"    }","highlight_start":1,"highlight_end":6},{"text":"}","highlight_start":1,"highlight_end":2}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null},{"message":"consider using a `let` binding to increase its lifetime","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":null}
{"message":"borrowed value does not live long enough (Ast)","code":{"code":"E0597","explanation":"\nThis error occurs because a borrow was made inside a variable which has a\ngreater lifetime than the borrowed one.\n\nExample of erroneous code:\n\n```compile_fail,E0597\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet mut x = Foo { x: None };\nlet y = 0;\nx.x = Some(&y); // error: `y` does not live long enough\n```\n\nIn here, `x` is created before `y` and therefore has a greater lifetime. Always\nkeep in mind that values in a scope are dropped in the opposite order they are\ncreated. So to fix the previous example, just make the `y` lifetime greater than\nthe `x`'s one:\n\n```\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet y = 0;\nlet mut x = Foo { x: None };\nx.x = Some(&y);\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":1114,"byte_end":1132,"line_start":38,"line_end":38,"column_start":17,"column_end":35,"is_primary":true,"text":[{"text":"        return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough","highlight_start":17,"highlight_end":35}],"label":"does not live long enough","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":1132,"byte_end":1133,"line_start":38,"line_end":38,"column_start":35,"column_end":36,"is_primary":false,"text":[{"text":"        return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough","highlight_start":35,"highlight_end":36}],"label":"temporary value only lives until here","suggested_replacement":null,"expansion":null}],"children":[{"message":"borrowed value must be valid for the lifetime 'a as defined on the function body at 30:1...","code":null,"level":"note","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":782,"byte_end":1201,"line_start":30,"line_end":41,"column_start":1,"column_end":2,"is_primary":true,"text":[{"text":"fn map_nums<'a,'b, F>(x: &ast, f: &mut F) -> &'a ast<'b> where F: FnMut(usize) -> usize {","highlight_start":1,"highlight_end":90},{"text":"    match *x {","highlight_start":1,"highlight_end":15},{"text":"      ast::num(x) => {","highlight_start":1,"highlight_end":23},{"text":"        return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough","highlight_start":1,"highlight_end":86},{"text":"      }","highlight_start":1,"highlight_end":8},{"text":"      ast::add(x, y) => {","highlight_start":1,"highlight_end":26},{"text":"        let m_x = map_nums(x, f);","highlight_start":1,"highlight_end":34},{"text":"        let m_y = map_nums(y, f);","highlight_start":1,"highlight_end":34},{"text":"        return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough","highlight_start":1,"highlight_end":88},{"text":"      }","highlight_start":1,"highlight_end":8},{"text":"    }","highlight_start":1,"highlight_end":6},{"text":"}","highlight_start":1,"highlight_end":2}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null},{"message":"consider using a `let` binding to increase its lifetime","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":null}
{"message":"cannot assign to `_` because it is borrowed (Mir)","code":{"code":"E0506","explanation":"\nThis error occurs when an attempt is made to assign to a borrowed value.\n\nExample of erroneous code:\n\n```compile_fail,E0506\nstruct FancyNum {\n    num: u8,\n}\n\nfn main() {\n    let mut fancy_num = FancyNum { num: 5 };\n    let fancy_ref = &fancy_num;\n    fancy_num = FancyNum { num: 6 };\n    // error: cannot assign to `fancy_num` because it is borrowed\n\n    println!(\"Num: {}, Ref: {}\", fancy_num.num, fancy_ref.num);\n}\n```\n\nBecause `fancy_ref` still holds a reference to `fancy_num`, `fancy_num` can't\nbe assigned to a new value as it would invalidate the reference.\n\nAlternatively, we can move out of `fancy_num` into a second `fancy_num`:\n\n```\nstruct FancyNum {\n    num: u8,\n}\n\nfn main() {\n    let mut fancy_num = FancyNum { num: 5 };\n    let moved_num = fancy_num;\n    fancy_num = FancyNum { num: 6 };\n\n    println!(\"Num: {}, Moved num: {}\", fancy_num.num, moved_num.num);\n}\n```\n\nIf the value has to be borrowed, try limiting the lifetime of the borrow using\na scoped block:\n\n```\nstruct FancyNum {\n    num: u8,\n}\n\nfn main() {\n    let mut fancy_num = FancyNum { num: 5 };\n\n    {\n        let fancy_ref = &fancy_num;\n        println!(\"Ref: {}\", fancy_ref.num);\n    }\n\n    // Works because `fancy_ref` is no longer in scope\n    fancy_num = FancyNum { num: 6 };\n    println!(\"Num: {}\", fancy_num.num);\n}\n```\n\nOr by moving the reference into a function:\n\n```\nstruct FancyNum {\n    num: u8,\n}\n\nfn main() {\n    let mut fancy_num = FancyNum { num: 5 };\n\n    print_fancy_ref(&fancy_num);\n\n    // Works because function borrow has ended\n    fancy_num = FancyNum { num: 6 };\n    println!(\"Num: {}\", fancy_num.num);\n}\n\nfn print_fancy_ref(fancy_ref: &FancyNum){\n    println!(\"Ref: {}\", fancy_ref.num);\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":925,"byte_end":943,"line_start":33,"line_end":33,"column_start":16,"column_end":34,"is_primary":false,"text":[{"text":"        return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough","highlight_start":16,"highlight_end":34}],"label":"borrow of `_` occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":944,"byte_end":944,"line_start":33,"line_end":33,"column_start":35,"column_end":35,"is_primary":true,"text":[{"text":"        return &ast::num((*f)(x)); //~ ERROR borrowed value does not live long enough","highlight_start":35,"highlight_end":35}],"label":"assignment to borrowed `_` occurs here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign to `_` because it is borrowed (Mir)","code":{"code":"E0506","explanation":"\nThis error occurs when an attempt is made to assign to a borrowed value.\n\nExample of erroneous code:\n\n```compile_fail,E0506\nstruct FancyNum {\n    num: u8,\n}\n\nfn main() {\n    let mut fancy_num = FancyNum { num: 5 };\n    let fancy_ref = &fancy_num;\n    fancy_num = FancyNum { num: 6 };\n    // error: cannot assign to `fancy_num` because it is borrowed\n\n    println!(\"Num: {}, Ref: {}\", fancy_num.num, fancy_ref.num);\n}\n```\n\nBecause `fancy_ref` still holds a reference to `fancy_num`, `fancy_num` can't\nbe assigned to a new value as it would invalidate the reference.\n\nAlternatively, we can move out of `fancy_num` into a second `fancy_num`:\n\n```\nstruct FancyNum {\n    num: u8,\n}\n\nfn main() {\n    let mut fancy_num = FancyNum { num: 5 };\n    let moved_num = fancy_num;\n    fancy_num = FancyNum { num: 6 };\n\n    println!(\"Num: {}, Moved num: {}\", fancy_num.num, moved_num.num);\n}\n```\n\nIf the value has to be borrowed, try limiting the lifetime of the borrow using\na scoped block:\n\n```\nstruct FancyNum {\n    num: u8,\n}\n\nfn main() {\n    let mut fancy_num = FancyNum { num: 5 };\n\n    {\n        let fancy_ref = &fancy_num;\n        println!(\"Ref: {}\", fancy_ref.num);\n    }\n\n    // Works because `fancy_ref` is no longer in scope\n    fancy_num = FancyNum { num: 6 };\n    println!(\"Num: {}\", fancy_num.num);\n}\n```\n\nOr by moving the reference into a function:\n\n```\nstruct FancyNum {\n    num: u8,\n}\n\nfn main() {\n    let mut fancy_num = FancyNum { num: 5 };\n\n    print_fancy_ref(&fancy_num);\n\n    // Works because function borrow has ended\n    fancy_num = FancyNum { num: 6 };\n    println!(\"Num: {}\", fancy_num.num);\n}\n\nfn print_fancy_ref(fancy_ref: &FancyNum){\n    println!(\"Ref: {}\", fancy_ref.num);\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":1113,"byte_end":1132,"line_start":38,"line_end":38,"column_start":16,"column_end":35,"is_primary":false,"text":[{"text":"        return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough","highlight_start":16,"highlight_end":35}],"label":"borrow of `_` occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/regions-creating-enums.rs","byte_start":1133,"byte_end":1133,"line_start":38,"line_end":38,"column_start":36,"column_end":36,"is_primary":true,"text":[{"text":"        return &ast::add(m_x, m_y);  //~ ERROR borrowed value does not live long enough","highlight_start":36,"highlight_end":36}],"label":"assignment to borrowed `_` occurs here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
