{"message":"cannot borrow immutable item `*x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/slice-mut-2.rs","byte_start":665,"byte_end":666,"line_start":17,"line_end":17,"column_start":18,"column_end":19,"is_primary":true,"text":[{"text":"    let _ = &mut x[2..4]; //~ERROR cannot borrow immutable borrowed content `*x` as mutable","highlight_start":18,"highlight_end":19}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `*x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/slice-mut-2.rs:17:18\n   |\n17 |     let _ = &mut x[2..4]; //~ERROR cannot borrow immutable borrowed content `*x` as mutable\n   |                  ^ cannot borrow as mutable\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
