{"message":"unnecessary lifetime parameter `'a`","code":null,"level":"warning","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/static-lifetime-bound.rs","byte_start":472,"byte_end":483,"line_start":11,"line_end":11,"column_start":6,"column_end":17,"is_primary":true,"text":[{"text":"fn f<'a: 'static>(_: &'a i32) {} //~WARN unnecessary lifetime parameter `'a`","highlight_start":6,"highlight_end":17}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"you can use the `'static` lifetime directly, in place of `'a`","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"warning: unnecessary lifetime parameter `'a`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/static-lifetime-bound.rs:11:6\n   |\n11 | fn f<'a: 'static>(_: &'a i32) {} //~WARN unnecessary lifetime parameter `'a`\n   |      ^^^^^^^^^^^\n   |\n   = help: you can use the `'static` lifetime directly, in place of `'a`\n\n"}
{"message":"borrowed value does not live long enough","code":{"code":"E0597","explanation":"\nThis error occurs because a borrow was made inside a variable which has a\ngreater lifetime than the borrowed one.\n\nExample of erroneous code:\n\n```compile_fail,E0597\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet mut x = Foo { x: None };\nlet y = 0;\nx.x = Some(&y); // error: `y` does not live long enough\n```\n\nIn here, `x` is created before `y` and therefore has a greater lifetime. Always\nkeep in mind that values in a scope are dropped in the opposite order they are\ncreated. So to fix the previous example, just make the `y` lifetime greater than\nthe `x`'s one:\n\n```\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet y = 0;\nlet mut x = Foo { x: None };\nx.x = Some(&y);\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/static-lifetime-bound.rs","byte_start":565,"byte_end":566,"line_start":14,"line_end":14,"column_start":9,"column_end":10,"is_primary":false,"text":[{"text":"    let x = 0;","highlight_start":9,"highlight_end":10}],"label":"temporary value created here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/static-lifetime-bound.rs","byte_start":618,"byte_end":619,"line_start":16,"line_end":16,"column_start":1,"column_end":2,"is_primary":true,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"temporary value dropped here while still borrowed","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/static-lifetime-bound.rs","byte_start":618,"byte_end":619,"line_start":16,"line_end":16,"column_start":1,"column_end":2,"is_primary":true,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"temporary value needs to live until here","suggested_replacement":null,"expansion":null}],"children":[{"message":"consider using a `let` binding to increase its lifetime","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0597]: borrowed value does not live long enough\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/static-lifetime-bound.rs:16:1\n   |\n14 |     let x = 0;\n   |         - temporary value created here\n15 |     f(&x); //~ERROR does not live long enough\n16 | }\n   | ^\n   | |\n   | temporary value dropped here while still borrowed\n   | temporary value needs to live until here\n   |\n   = note: consider using a `let` binding to increase its lifetime\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
