{"message":"cannot borrow immutable item `tick1` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs","byte_start":946,"byte_end":1041,"line_start":26,"line_end":28,"column_start":17,"column_end":6,"is_primary":true,"text":[{"text":"    let tick2 = || { //~ ERROR closure cannot assign to immutable local variable `tick1`","highlight_start":17,"highlight_end":89},{"text":"        tick1();","highlight_start":1,"highlight_end":17},{"text":"    };","highlight_start":1,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `tick1` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs:26:17\n   |\n26 |       let tick2 = || { //~ ERROR closure cannot assign to immutable local variable `tick1`\n   |  _________________^\n27 | |         tick1();\n28 | |     };\n   | |_____^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `tick2` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs","byte_start":1048,"byte_end":1053,"line_start":30,"line_end":30,"column_start":5,"column_end":10,"is_primary":true,"text":[{"text":"    tick2(); //~ ERROR cannot borrow","highlight_start":5,"highlight_end":10}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `tick2` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs:30:5\n   |\n30 |     tick2(); //~ ERROR cannot borrow\n   |     ^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `tick1` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs","byte_start":1027,"byte_end":1032,"line_start":27,"line_end":27,"column_start":9,"column_end":14,"is_primary":true,"text":[{"text":"        tick1();","highlight_start":9,"highlight_end":14}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `tick1`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `tick1` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs:27:9\n   |\n27 |         tick1();\n   |         ^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `tick1`\n\n"}
{"message":"aborting due to 3 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 3 previous errors\n\n"}
