{"message":"closure cannot assign to immutable local variable `tick1` (Ast)","code":{"code":"E0595","explanation":"\nClosures cannot mutate immutable captured variables.\n\nErroneous code example:\n\n```compile_fail,E0595\nlet x = 3; // error: closure cannot assign to immutable local variable `x`\nlet mut c = || { x += 1 };\n```\n\nMake the variable binding mutable:\n\n```\nlet mut x = 3; // ok!\nlet mut c = || { x += 1 };\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs","byte_start":946,"byte_end":948,"line_start":26,"line_end":26,"column_start":17,"column_end":19,"is_primary":true,"text":[{"text":"    let tick2 = || { //~ ERROR closure cannot assign to immutable local variable `tick1`","highlight_start":17,"highlight_end":19}],"label":"cannot borrow mutably","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs","byte_start":719,"byte_end":724,"line_start":19,"line_end":19,"column_start":9,"column_end":14,"is_primary":false,"text":[{"text":"    let tick1 = || {","highlight_start":9,"highlight_end":14}],"label":"consider changing this to `mut tick1`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable local variable `tick2` as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs","byte_start":1048,"byte_end":1053,"line_start":30,"line_end":30,"column_start":5,"column_end":10,"is_primary":true,"text":[{"text":"    tick2(); //~ ERROR cannot borrow","highlight_start":5,"highlight_end":10}],"label":"cannot borrow mutably","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/unboxed-closures-infer-fnmut-calling-fnmut-no-mut.rs","byte_start":938,"byte_end":943,"line_start":26,"line_end":26,"column_start":9,"column_end":14,"is_primary":false,"text":[{"text":"    let tick2 = || { //~ ERROR closure cannot assign to immutable local variable `tick1`","highlight_start":9,"highlight_end":14}],"label":"consider changing this to `mut tick2`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
