{"message":"cannot borrow immutable item `tick` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-infer-fnmut-missing-mut.rs","byte_start":642,"byte_end":646,"line_start":17,"line_end":17,"column_start":5,"column_end":9,"is_primary":true,"text":[{"text":"    tick(); //~ ERROR cannot borrow immutable local variable `tick` as mutable","highlight_start":5,"highlight_end":9}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `tick` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-infer-fnmut-missing-mut.rs:17:5\n   |\n17 |     tick(); //~ ERROR cannot borrow immutable local variable `tick` as mutable\n   |     ^^^^ cannot borrow as mutable\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
