{"message":"cannot borrow immutable item `n` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs","byte_start":837,"byte_end":895,"line_start":24,"line_end":26,"column_start":27,"column_end":6,"is_primary":true,"text":[{"text":"    let mut f = to_fn_mut(|| { //~ ERROR closure cannot assign","highlight_start":27,"highlight_end":63},{"text":"        n += 1;","highlight_start":1,"highlight_end":16},{"text":"    });","highlight_start":1,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `n` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs:24:27\n   |\n24 |       let mut f = to_fn_mut(|| { //~ ERROR closure cannot assign\n   |  ___________________________^\n25 | |         n += 1;\n26 | |     });\n   | |_____^ cannot borrow as mutable\n\n"}
{"message":"cannot assign to immutable item `n`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs","byte_start":882,"byte_end":888,"line_start":25,"line_end":25,"column_start":9,"column_end":15,"is_primary":true,"text":[{"text":"        n += 1;","highlight_start":9,"highlight_end":15}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `n`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `n`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs:25:9\n   |\n25 |         n += 1;\n   |         ^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `n`\n\n"}
{"message":"cannot assign to immutable item `n`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs","byte_start":1226,"byte_end":1232,"line_start":42,"line_end":42,"column_start":9,"column_end":15,"is_primary":true,"text":[{"text":"        n += 1; //~ ERROR cannot assign","highlight_start":9,"highlight_end":15}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `n`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs:42:9\n   |\n42 |         n += 1; //~ ERROR cannot assign\n   |         ^^^^^^ cannot mutate\n\n"}
{"message":"cannot assign to immutable item `n`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs","byte_start":1430,"byte_end":1436,"line_start":56,"line_end":56,"column_start":9,"column_end":15,"is_primary":true,"text":[{"text":"        n += 1; //~ ERROR cannot assign","highlight_start":9,"highlight_end":15}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `n`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs:56:9\n   |\n56 |         n += 1; //~ ERROR cannot assign\n   |         ^^^^^^ cannot mutate\n\n"}
{"message":"cannot assign to immutable item `n`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs","byte_start":1541,"byte_end":1547,"line_start":63,"line_end":63,"column_start":9,"column_end":15,"is_primary":true,"text":[{"text":"        n += 1; //~ ERROR cannot assign","highlight_start":9,"highlight_end":15}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `n`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closures-mutate-upvar.rs:63:9\n   |\n63 |         n += 1; //~ ERROR cannot assign\n   |         ^^^^^^ cannot mutate\n\n"}
{"message":"aborting due to 5 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 5 previous errors\n\n"}
