{"message":"cannot borrow immutable item `v` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/writing-to-immutable-vec.rs","byte_start":523,"byte_end":524,"line_start":14,"line_end":14,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    v[1] = 4; //~ ERROR cannot borrow immutable local variable `v` as mutable","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `v` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/writing-to-immutable-vec.rs:14:5\n   |\n14 |     v[1] = 4; //~ ERROR cannot borrow immutable local variable `v` as mutable\n   |     ^ cannot borrow as mutable\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
