{"message":"cannot borrow immutable item `x.0` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-ref-mut-of-imm.rs","byte_start":553,"byte_end":562,"line_start":14,"line_end":14,"column_start":12,"column_end":21,"is_primary":true,"text":[{"text":"      Some(ref mut v) => *v //~ ERROR cannot borrow","highlight_start":12,"highlight_end":21}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `x`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `x.0` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-ref-mut-of-imm.rs:14:12\n   |\n14 |       Some(ref mut v) => *v //~ ERROR cannot borrow\n   |            ^^^^^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `x`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
