{"message":"use of possibly uninitialized variable: `s.a` (Ast)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-union-uninitialized.rs","byte_start":631,"byte_end":633,"line_start":25,"line_end":25,"column_start":13,"column_end":15,"is_primary":true,"text":[{"text":"        let sa = s.a; //~ ERROR use of possibly uninitialized variable: `s.a`","highlight_start":13,"highlight_end":15}],"label":"use of possibly uninitialized `s.a`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"use of possibly uninitialized variable: `u.a` (Ast)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-union-uninitialized.rs","byte_start":709,"byte_end":711,"line_start":26,"line_end":26,"column_start":13,"column_end":15,"is_primary":true,"text":[{"text":"        let ua = u.a; //~ ERROR use of possibly uninitialized variable: `u.a`","highlight_start":13,"highlight_end":15}],"label":"use of possibly uninitialized `u.a`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"use of possibly uninitialized variable: `u.a` (Mir)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-union-uninitialized.rs","byte_start":714,"byte_end":717,"line_start":26,"line_end":26,"column_start":18,"column_end":21,"is_primary":true,"text":[{"text":"        let ua = u.a; //~ ERROR use of possibly uninitialized variable: `u.a`","highlight_start":18,"highlight_end":21}],"label":"use of possibly uninitialized `u.a`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 3 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
