{"message":"cannot borrow immutable field as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1213,"byte_end":1216,"line_start":57,"line_end":57,"column_start":24,"column_end":27,"is_primary":true,"text":[{"text":"    let __isize = &mut x.y; //~ ERROR cannot borrow","highlight_start":24,"highlight_end":27}],"label":"cannot mutably borrow immutable field","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable field as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1308,"byte_end":1311,"line_start":61,"line_end":61,"column_start":24,"column_end":27,"is_primary":true,"text":[{"text":"    let __isize = &mut x.y; //~ ERROR cannot borrow","highlight_start":24,"highlight_end":27}],"label":"cannot mutably borrow immutable field","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable field as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1468,"byte_end":1471,"line_start":69,"line_end":69,"column_start":10,"column_end":13,"is_primary":true,"text":[{"text":"    &mut x.y //~ ERROR cannot borrow","highlight_start":10,"highlight_end":13}],"label":"cannot mutably borrow immutable field","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable field as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1570,"byte_end":1573,"line_start":73,"line_end":73,"column_start":10,"column_end":13,"is_primary":true,"text":[{"text":"    &mut x.y //~ ERROR cannot borrow","highlight_start":10,"highlight_end":13}],"label":"cannot mutably borrow immutable field","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign to immutable field (Ast)","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1642,"byte_end":1649,"line_start":77,"line_end":77,"column_start":5,"column_end":12,"is_primary":true,"text":[{"text":"    x.y = 3; //~ ERROR cannot assign","highlight_start":5,"highlight_end":12}],"label":"cannot mutably borrow immutable field","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign to immutable field (Ast)","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1723,"byte_end":1730,"line_start":81,"line_end":81,"column_start":5,"column_end":12,"is_primary":true,"text":[{"text":"    x.y = 3; //~ ERROR cannot assign","highlight_start":5,"highlight_end":12}],"label":"cannot mutably borrow immutable field","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign to immutable field (Ast)","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1808,"byte_end":1815,"line_start":85,"line_end":85,"column_start":5,"column_end":12,"is_primary":true,"text":[{"text":"    x.y = 3; //~ ERROR cannot assign","highlight_start":5,"highlight_end":12}],"label":"cannot mutably borrow immutable field","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable borrowed content as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1951,"byte_end":1952,"line_start":93,"line_end":93,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    x.set(0, 0); //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable borrowed content as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2036,"byte_end":2037,"line_start":97,"line_end":97,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    x.set(0, 0); //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable borrowed content as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2206,"byte_end":2207,"line_start":105,"line_end":105,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    x.y_mut() //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable borrowed content as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2310,"byte_end":2311,"line_start":109,"line_end":109,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    x.y_mut() //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable borrowed content as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2390,"byte_end":2391,"line_start":113,"line_end":113,"column_start":6,"column_end":7,"is_primary":true,"text":[{"text":"    *x.y_mut() = 3; //~ ERROR cannot borrow","highlight_start":6,"highlight_end":7}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable borrowed content as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2479,"byte_end":2480,"line_start":117,"line_end":117,"column_start":6,"column_end":7,"is_primary":true,"text":[{"text":"    *x.y_mut() = 3; //~ ERROR cannot borrow","highlight_start":6,"highlight_end":7}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable borrowed content as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2572,"byte_end":2573,"line_start":121,"line_end":121,"column_start":6,"column_end":7,"is_primary":true,"text":[{"text":"    *x.y_mut() = 3; //~ ERROR cannot borrow","highlight_start":6,"highlight_end":7}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 14 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
