{"message":"cannot borrow immutable argument `arg` as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":568,"byte_end":571,"line_start":20,"line_end":20,"column_start":5,"column_end":8,"is_primary":true,"text":[{"text":"    arg.mutate(); //~ ERROR: cannot borrow immutable argument","highlight_start":5,"highlight_end":8}],"label":"cannot borrow mutably","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":554,"byte_end":557,"line_start":19,"line_end":19,"column_start":9,"column_end":12,"is_primary":false,"text":[{"text":"fn func(arg: S) {","highlight_start":9,"highlight_end":12}],"label":"consider changing this to `mut arg`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable argument `arg` as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":677,"byte_end":680,"line_start":25,"line_end":25,"column_start":9,"column_end":12,"is_primary":true,"text":[{"text":"        arg.mutate(); //~ ERROR: cannot borrow immutable argument","highlight_start":9,"highlight_end":12}],"label":"cannot borrow mutably","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":659,"byte_end":662,"line_start":24,"line_end":24,"column_start":22,"column_end":25,"is_primary":false,"text":[{"text":"    fn method(&self, arg: S) {","highlight_start":22,"highlight_end":25}],"label":"consider changing this to `mut arg`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable argument `arg` as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":794,"byte_end":797,"line_start":31,"line_end":31,"column_start":9,"column_end":12,"is_primary":true,"text":[{"text":"        arg.mutate(); //~ ERROR: cannot borrow immutable argument","highlight_start":9,"highlight_end":12}],"label":"cannot borrow mutably","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":776,"byte_end":779,"line_start":30,"line_end":30,"column_start":23,"column_end":26,"is_primary":false,"text":[{"text":"    fn default(&self, arg: S) {","highlight_start":23,"highlight_end":26}],"label":"consider changing this to `mut arg`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable argument `arg` as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":969,"byte_end":972,"line_start":42,"line_end":42,"column_start":17,"column_end":20,"is_primary":true,"text":[{"text":"    (|arg: S| { arg.mutate() })(s); //~ ERROR: cannot borrow immutable argument","highlight_start":17,"highlight_end":20}],"label":"cannot borrow mutably","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":959,"byte_end":962,"line_start":42,"line_end":42,"column_start":7,"column_end":10,"is_primary":false,"text":[{"text":"    (|arg: S| { arg.mutate() })(s); //~ ERROR: cannot borrow immutable argument","highlight_start":7,"highlight_end":10}],"label":"consider changing this to `mut arg`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
