{"message":"cannot assign to immutable item `x`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs","byte_start":801,"byte_end":807,"line_start":25,"line_end":25,"column_start":27,"column_end":33,"is_primary":true,"text":[{"text":"        let _f = to_fn(|| x = 42); //~ ERROR cannot assign","highlight_start":27,"highlight_end":33}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs:25:27\n   |\n25 |         let _f = to_fn(|| x = 42); //~ ERROR cannot assign\n   |                           ^^^^^^ cannot mutate\n\n"}
{"message":"cannot borrow immutable item `y` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs","byte_start":888,"byte_end":894,"line_start":28,"line_end":28,"column_start":31,"column_end":37,"is_primary":true,"text":[{"text":"        let _g = to_fn(|| set(&mut y)); //~ ERROR cannot borrow","highlight_start":31,"highlight_end":37}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `y` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs:28:31\n   |\n28 |         let _g = to_fn(|| set(&mut y)); //~ ERROR cannot borrow\n   |                               ^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot assign to immutable item `z`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs","byte_start":1000,"byte_end":1006,"line_start":31,"line_end":31,"column_start":55,"column_end":61,"is_primary":true,"text":[{"text":"        let _h = to_fn_mut(|| { set(&mut z); to_fn(|| z = 42); }); //~ ERROR cannot assign","highlight_start":55,"highlight_end":61}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `z`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs:31:55\n   |\n31 |         let _h = to_fn_mut(|| { set(&mut z); to_fn(|| z = 42); }); //~ ERROR cannot assign\n   |                                                       ^^^^^^ cannot mutate\n\n"}
{"message":"cannot assign to immutable item `x`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs","byte_start":1129,"byte_end":1135,"line_start":37,"line_end":37,"column_start":32,"column_end":38,"is_primary":true,"text":[{"text":"        let _f = to_fn(move || x = 42); //~ ERROR cannot assign","highlight_start":32,"highlight_end":38}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs:37:32\n   |\n37 |         let _f = to_fn(move || x = 42); //~ ERROR cannot assign\n   |                                ^^^^^^ cannot mutate\n\n"}
{"message":"cannot borrow immutable item `y` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs","byte_start":1221,"byte_end":1227,"line_start":40,"line_end":40,"column_start":36,"column_end":42,"is_primary":true,"text":[{"text":"        let _g = to_fn(move || set(&mut y)); //~ ERROR cannot borrow","highlight_start":36,"highlight_end":42}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `y` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs:40:36\n   |\n40 |         let _g = to_fn(move || set(&mut y)); //~ ERROR cannot borrow\n   |                                    ^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot assign to immutable item `z`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs","byte_start":1343,"byte_end":1349,"line_start":43,"line_end":43,"column_start":65,"column_end":71,"is_primary":true,"text":[{"text":"        let _h = to_fn_mut(move || { set(&mut z); to_fn(move || z = 42); }); //~ ERROR cannot assign","highlight_start":65,"highlight_end":71}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `z`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrow-immutable-upvar-mutation.rs:43:65\n   |\n43 |         let _h = to_fn_mut(move || { set(&mut z); to_fn(move || z = 42); }); //~ ERROR cannot assign\n   |                                                                 ^^^^^^ cannot mutate\n\n"}
{"message":"aborting due to 6 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 6 previous errors\n\n"}
