{"message":"references in constants may only refer to immutable values","code":{"code":"E0017","explanation":"\nReferences in statics and constants may only refer to immutable values.\nErroneous code example:\n\n```compile_fail,E0017\nstatic X: i32 = 1;\nconst C: i32 = 2;\n\n// these three are not allowed:\nconst CR: &'static mut i32 = &mut C;\nstatic STATIC_REF: &'static mut i32 = &mut X;\nstatic CONST_REF: &'static mut i32 = &mut C;\n```\n\nStatics are shared everywhere, and if they refer to mutable data one might\nviolate memory safety since holding multiple mutable references to shared data\nis not allowed.\n\nIf you really want global mutable state, try using `static mut` or a global\n`UnsafeCell`.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/E0017.rs","byte_start":534,"byte_end":540,"line_start":14,"line_end":14,"column_start":30,"column_end":36,"is_primary":true,"text":[{"text":"const CR: &'static mut i32 = &mut C; //~ ERROR E0017","highlight_start":30,"highlight_end":36}],"label":"constants require immutable values","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"references in statics may only refer to immutable values","code":{"code":"E0017","explanation":"\nReferences in statics and constants may only refer to immutable values.\nErroneous code example:\n\n```compile_fail,E0017\nstatic X: i32 = 1;\nconst C: i32 = 2;\n\n// these three are not allowed:\nconst CR: &'static mut i32 = &mut C;\nstatic STATIC_REF: &'static mut i32 = &mut X;\nstatic CONST_REF: &'static mut i32 = &mut C;\n```\n\nStatics are shared everywhere, and if they refer to mutable data one might\nviolate memory safety since holding multiple mutable references to shared data\nis not allowed.\n\nIf you really want global mutable state, try using `static mut` or a global\n`UnsafeCell`.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/E0017.rs","byte_start":678,"byte_end":684,"line_start":16,"line_end":16,"column_start":39,"column_end":45,"is_primary":true,"text":[{"text":"static STATIC_REF: &'static mut i32 = &mut X; //~ ERROR E0017","highlight_start":39,"highlight_end":45}],"label":"statics require immutable values","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable static item as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/E0017.rs","byte_start":683,"byte_end":684,"line_start":16,"line_end":16,"column_start":44,"column_end":45,"is_primary":true,"text":[{"text":"static STATIC_REF: &'static mut i32 = &mut X; //~ ERROR E0017","highlight_start":44,"highlight_end":45}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"references in statics may only refer to immutable values","code":{"code":"E0017","explanation":"\nReferences in statics and constants may only refer to immutable values.\nErroneous code example:\n\n```compile_fail,E0017\nstatic X: i32 = 1;\nconst C: i32 = 2;\n\n// these three are not allowed:\nconst CR: &'static mut i32 = &mut C;\nstatic STATIC_REF: &'static mut i32 = &mut X;\nstatic CONST_REF: &'static mut i32 = &mut C;\n```\n\nStatics are shared everywhere, and if they refer to mutable data one might\nviolate memory safety since holding multiple mutable references to shared data\nis not allowed.\n\nIf you really want global mutable state, try using `static mut` or a global\n`UnsafeCell`.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/E0017.rs","byte_start":899,"byte_end":905,"line_start":19,"line_end":19,"column_start":38,"column_end":44,"is_primary":true,"text":[{"text":"static CONST_REF: &'static mut i32 = &mut C; //~ ERROR E0017","highlight_start":38,"highlight_end":44}],"label":"statics require immutable values","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 7 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
