{"message":"cannot borrow immutable indexed content as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/issue-41726.rs","byte_start":612,"byte_end":632,"line_start":15,"line_end":15,"column_start":9,"column_end":29,"is_primary":true,"text":[{"text":"        things[src.as_str()].sort(); //~ ERROR cannot borrow immutable","highlight_start":9,"highlight_end":29}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign twice to immutable variable `val` (Mir)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/issue-41726.rs","byte_start":577,"byte_end":680,"line_start":14,"line_end":16,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    for src in things.keys() {","highlight_start":5,"highlight_end":31},{"text":"        things[src.as_str()].sort(); //~ ERROR cannot borrow immutable","highlight_start":1,"highlight_end":71},{"text":"    }","highlight_start":1,"highlight_end":6}],"label":"cannot assign twice to immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/issue-41726.rs","byte_start":577,"byte_end":680,"line_start":14,"line_end":16,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    for src in things.keys() {","highlight_start":5,"highlight_end":31},{"text":"        things[src.as_str()].sort(); //~ ERROR cannot borrow immutable","highlight_start":1,"highlight_end":71},{"text":"    }","highlight_start":1,"highlight_end":6}],"label":"first assignment to `val`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign twice to immutable variable `src` (Mir)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/issue-41726.rs","byte_start":577,"byte_end":680,"line_start":14,"line_end":16,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    for src in things.keys() {","highlight_start":5,"highlight_end":31},{"text":"        things[src.as_str()].sort(); //~ ERROR cannot borrow immutable","highlight_start":1,"highlight_end":71},{"text":"    }","highlight_start":1,"highlight_end":6}],"label":"cannot assign twice to immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/issue-41726.rs","byte_start":577,"byte_end":680,"line_start":14,"line_end":16,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    for src in things.keys() {","highlight_start":5,"highlight_end":31},{"text":"        things[src.as_str()].sort(); //~ ERROR cannot borrow immutable","highlight_start":1,"highlight_end":71},{"text":"    }","highlight_start":1,"highlight_end":6}],"label":"first assignment to `src`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 3 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
