{"message":"cannot borrow data mutably in a captured outer variable in an `Fn` closure (Ast)","code":{"code":"E0387","explanation":"\nThis error occurs when an attempt is made to mutate or mutably reference data\nthat a closure has captured immutably. Examples of this error are shown below:\n\n```compile_fail,E0387\n// Accepts a function or a closure that captures its environment immutably.\n// Closures passed to foo will not be able to mutate their closed-over state.\nfn foo<F: Fn()>(f: F) { }\n\n// Attempts to mutate closed-over data. Error message reads:\n// `cannot assign to data in a captured outer variable...`\nfn mutable() {\n    let mut x = 0u32;\n    foo(|| x = 2);\n}\n\n// Attempts to take a mutable reference to closed-over data.  Error message\n// reads: `cannot borrow data mutably in a captured outer variable...`\nfn mut_addr() {\n    let mut x = 0u32;\n    foo(|| { let y = &mut x; });\n}\n```\n\nThe problem here is that foo is defined as accepting a parameter of type `Fn`.\nClosures passed into foo will thus be inferred to be of type `Fn`, meaning that\nthey capture their context immutably.\n\nIf the definition of `foo` is under your control, the simplest solution is to\ncapture the data mutably. This can be done by defining `foo` to take FnMut\nrather than Fn:\n\n```\nfn foo<F: FnMut()>(f: F) { }\n```\n\nAlternatively, we can consider using the `Cell` and `RefCell` types to achieve\ninterior mutability through a shared reference. Our example's `mutable`\nfunction could be redefined as below:\n\n```\nuse std::cell::Cell;\n\nfn foo<F: Fn()>(f: F) { }\n\nfn mutable() {\n    let x = Cell::new(0u32);\n    foo(|| x.set(2));\n}\n```\n\nYou can read more about cell types in the API documentation:\n\nhttps://doc.rust-lang.org/std/cell/\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-21600.rs","byte_start":754,"byte_end":756,"line_start":24,"line_end":24,"column_start":17,"column_end":19,"is_primary":true,"text":[{"text":"        call_it(|| x.gen_mut()); //~ ERROR cannot borrow data mutably in a captured outer","highlight_start":17,"highlight_end":19}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"consider changing this to accept closures that implement `FnMut`","code":null,"level":"help","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-21600.rs","byte_start":627,"byte_end":982,"line_start":22,"line_end":27,"column_start":13,"column_end":6,"is_primary":true,"text":[{"text":"    call_it(|| {    //~ HELP consider changing this to accept closures that implement `FnMut`","highlight_start":13,"highlight_end":94},{"text":"        call_it(|| x.gen());","highlight_start":1,"highlight_end":29},{"text":"        call_it(|| x.gen_mut()); //~ ERROR cannot borrow data mutably in a captured outer","highlight_start":1,"highlight_end":90},{"text":"        //~^ ERROR cannot borrow data mutably in a captured outer","highlight_start":1,"highlight_end":66},{"text":"        //~| HELP consider changing this closure to take self by mutable reference","highlight_start":1,"highlight_end":83},{"text":"    });","highlight_start":1,"highlight_end":6}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}],"rendered":null}
{"message":"cannot borrow data mutably in a captured outer variable in an `Fn` closure (Ast)","code":{"code":"E0387","explanation":"\nThis error occurs when an attempt is made to mutate or mutably reference data\nthat a closure has captured immutably. Examples of this error are shown below:\n\n```compile_fail,E0387\n// Accepts a function or a closure that captures its environment immutably.\n// Closures passed to foo will not be able to mutate their closed-over state.\nfn foo<F: Fn()>(f: F) { }\n\n// Attempts to mutate closed-over data. Error message reads:\n// `cannot assign to data in a captured outer variable...`\nfn mutable() {\n    let mut x = 0u32;\n    foo(|| x = 2);\n}\n\n// Attempts to take a mutable reference to closed-over data.  Error message\n// reads: `cannot borrow data mutably in a captured outer variable...`\nfn mut_addr() {\n    let mut x = 0u32;\n    foo(|| { let y = &mut x; });\n}\n```\n\nThe problem here is that foo is defined as accepting a parameter of type `Fn`.\nClosures passed into foo will thus be inferred to be of type `Fn`, meaning that\nthey capture their context immutably.\n\nIf the definition of `foo` is under your control, the simplest solution is to\ncapture the data mutably. This can be done by defining `foo` to take FnMut\nrather than Fn:\n\n```\nfn foo<F: FnMut()>(f: F) { }\n```\n\nAlternatively, we can consider using the `Cell` and `RefCell` types to achieve\ninterior mutability through a shared reference. Our example's `mutable`\nfunction could be redefined as below:\n\n```\nuse std::cell::Cell;\n\nfn foo<F: Fn()>(f: F) { }\n\nfn mutable() {\n    let x = Cell::new(0u32);\n    foo(|| x.set(2));\n}\n```\n\nYou can read more about cell types in the API documentation:\n\nhttps://doc.rust-lang.org/std/cell/\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-21600.rs","byte_start":757,"byte_end":758,"line_start":24,"line_end":24,"column_start":20,"column_end":21,"is_primary":true,"text":[{"text":"        call_it(|| x.gen_mut()); //~ ERROR cannot borrow data mutably in a captured outer","highlight_start":20,"highlight_end":21}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"consider changing this closure to take self by mutable reference","code":null,"level":"help","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/issue-21600.rs","byte_start":754,"byte_end":768,"line_start":24,"line_end":24,"column_start":17,"column_end":31,"is_primary":true,"text":[{"text":"        call_it(|| x.gen_mut()); //~ ERROR cannot borrow data mutably in a captured outer","highlight_start":17,"highlight_end":31}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}],"rendered":null}
{"message":"/home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:1607: End-user description not implemented for field access on `TyClosure(DefId { krate: CrateNum(0), index: DefIndex(1:11) => issue_21600[317d]::main[0]::{{closure}}[0] }, ClosureSubsts { substs: Slice([&mut A]) })`","code":null,"level":"error: internal compiler error","spans":[],"children":[],"rendered":null}
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.23.0-dev running on x86_64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', /home/ariel/Rust/rust-master/src/librustc_errors/lib.rs:485:8
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /home/ariel/Rust/rust-master/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /home/ariel/Rust/rust-master/src/libstd/sys_common/backtrace.rs:68
   2: std::panicking::default_hook::{{closure}}
             at /home/ariel/Rust/rust-master/src/libstd/sys_common/backtrace.rs:57
             at /home/ariel/Rust/rust-master/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /home/ariel/Rust/rust-master/src/libstd/panicking.rs:391
   4: std::panicking::rust_panic_with_hook
             at /home/ariel/Rust/rust-master/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /home/ariel/Rust/rust-master/src/libstd/panicking.rs:538
   6: rustc_errors::Handler::bug
             at /home/ariel/Rust/rust-master/src/librustc_errors/lib.rs:485
   7: rustc::session::opt_span_bug_fmt::{{closure}}
             at /home/ariel/Rust/rust-master/src/librustc/session/mod.rs:992
   8: rustc::session::opt_span_bug_fmt
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:0
   9: rustc::session::bug_fmt
             at /home/ariel/Rust/rust-master/src/librustc/session/mod.rs:972
  10: rustc_mir::borrow_check::MirBorrowckCtxt::describe_field_from_ty
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:1607
  11: rustc_mir::borrow_check::MirBorrowckCtxt::describe_field_from_ty
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:0
  12: rustc_mir::borrow_check::MirBorrowckCtxt::describe_field
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:0
  13: rustc_mir::borrow_check::MirBorrowckCtxt::describe_field
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:1565
  14: rustc_mir::borrow_check::MirBorrowckCtxt::append_lvalue_to_string
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:1515
  15: rustc_mir::borrow_check::MirBorrowckCtxt::append_lvalue_to_string
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:1530
  16: rustc_mir::borrow_check::MirBorrowckCtxt::access_lvalue
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:1488
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:877
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:450
  17: <rustc_mir::borrow_check::MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx> as rustc_mir::dataflow::DataflowResultsConsumer<'b, 'tcx>>::visit_statement_entry
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:533
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:245
  18: rustc_mir::borrow_check::do_mir_borrowck
             at /home/ariel/Rust/rust-master/src/librustc_mir/dataflow/mod.rs:309
             at /home/ariel/Rust/rust-master/src/librustc_mir/dataflow/mod.rs:299
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:147
  19: rustc_mir::borrow_check::mir_borrowck
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:64
             at /home/ariel/Rust/rust-master/src/librustc/infer/mod.rs:375
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:1522
             at /home/ariel/Rust/rust-master/src/libstd/thread/local.rs:377
             at /home/ariel/Rust/rust-master/src/libstd/thread/local.rs:288
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:1519
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:1330
             at /home/ariel/Rust/rust-master/src/librustc/infer/mod.rs:375
             at /home/ariel/Rust/rust-master/src/librustc_mir/borrow_check.rs:62
  20: rustc::dep_graph::graph::DepGraph::with_task_impl
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:370
             at /home/ariel/Rust/rust-master/src/librustc/dep_graph/graph.rs:273
  21: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_borrowck<'tcx>>::force
             at /home/ariel/Rust/rust-master/src/librustc/dep_graph/graph.rs:189
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:452
             at /home/ariel/Rust/rust-master/src/librustc_errors/lib.rs:548
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:445
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:115
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:444
  22: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_borrowck<'tcx>>::try_get
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:288
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:486
  23: rustc::ty::maps::TyCtxtAt::mir_borrowck
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:525
  24: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::mir_borrowck
             at /home/ariel/Rust/rust-master/src/librustc/ty/maps/plumbing.rs:518
  25: rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::{{closure}}
             at /home/ariel/Rust/rust-master/src/librustc_driver/driver.rs:1064
  26: rustc::ty::context::TyCtxt::create_and_enter
             at /home/ariel/Rust/rust-master/src/librustc/util/common.rs:120
             at /home/ariel/Rust/rust-master/src/librustc_driver/driver.rs:1062
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:1522
             at /home/ariel/Rust/rust-master/src/libstd/thread/local.rs:377
             at /home/ariel/Rust/rust-master/src/libstd/thread/local.rs:288
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:1519
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:1506
             at /home/ariel/Rust/rust-master/src/libstd/thread/local.rs:377
             at /home/ariel/Rust/rust-master/src/libstd/thread/local.rs:288
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:1503
             at /home/ariel/Rust/rust-master/src/librustc/ty/context.rs:1138
  27: rustc_driver::driver::compile_input
             at /home/ariel/Rust/rust-master/src/librustc_driver/driver.rs:1010
             at /home/ariel/Rust/rust-master/src/librustc_driver/driver.rs:209
  28: rustc_driver::run_compiler
             at /home/ariel/Rust/rust-master/src/librustc_driver/lib.rs:253

