{"message":"cannot move out of type `[D; 4]`, a non-copy array (Ast)","code":{"code":"E0508","explanation":"\nA value was moved out of a non-copy fixed-size array.\n\nExample of erroneous code:\n\n```compile_fail,E0508\nstruct NonCopy;\n\nfn main() {\n    let array = [NonCopy; 1];\n    let _value = array[0]; // error: cannot move out of type `[NonCopy; 1]`,\n                           //        a non-copy fixed-size array\n}\n```\n\nThe first element was moved out of the array, but this is not\npossible because `NonCopy` does not implement the `Copy` trait.\n\nConsider borrowing the element instead of moving it:\n\n```\nstruct NonCopy;\n\nfn main() {\n    let array = [NonCopy; 1];\n    let _value = &array[0]; // Borrowing is allowed, unlike moving.\n}\n```\n\nAlternatively, if your type implements `Clone` and you need to own the value,\nconsider borrowing and then cloning:\n\n```\n#[derive(Clone)]\nstruct NonCopy;\n\nfn main() {\n    let array = [NonCopy; 1];\n    // Now you can clone the array element.\n    let _value = array[0].clone();\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/move-out-of-array-1.rs","byte_start":821,"byte_end":825,"line_start":27,"line_end":27,"column_start":5,"column_end":9,"is_primary":true,"text":[{"text":"    a[i] //~ ERROR cannot move out of type `[D; 4]`, a non-copy array","highlight_start":5,"highlight_end":9}],"label":"cannot move out of here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"/home/ariel/Rust/rust-master/src/librustc_mir/util/borrowck_errors.rs:274: this path should not cause illegal move","code":null,"level":"error: internal compiler error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/move-out-of-array-1.rs","byte_start":821,"byte_end":825,"line_start":27,"line_end":27,"column_start":5,"column_end":9,"is_primary":true,"text":[{"text":"    a[i] //~ ERROR cannot move out of type `[D; 4]`, a non-copy array","highlight_start":5,"highlight_end":9}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.23.0-dev running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'Box<Any>', /home/ariel/Rust/rust-master/src/librustc_errors/lib.rs:450:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.

