{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":884,"byte_end":892,"line_start":23,"line_end":23,"column_start":5,"column_end":13,"is_primary":true,"text":[{"text":"    || x = 1; //~ ERROR cannot assign","highlight_start":5,"highlight_end":13}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:23:5\n   |\n23 |     || x = 1; //~ ERROR cannot assign\n   |     ^^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":982,"byte_end":996,"line_start":25,"line_end":25,"column_start":5,"column_end":19,"is_primary":true,"text":[{"text":"    || set(&mut x); //~ ERROR cannot assign","highlight_start":5,"highlight_end":19}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:25:5\n   |\n25 |     || set(&mut x); //~ ERROR cannot assign\n   |     ^^^^^^^^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":1026,"byte_end":1034,"line_start":26,"line_end":26,"column_start":5,"column_end":13,"is_primary":true,"text":[{"text":"    || x = 1; //~ ERROR cannot assign","highlight_start":5,"highlight_end":13}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:26:5\n   |\n26 |     || x = 1; //~ ERROR cannot assign\n   |     ^^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":1124,"byte_end":1138,"line_start":28,"line_end":28,"column_start":5,"column_end":19,"is_primary":true,"text":[{"text":"    || set(&mut x); //~ ERROR cannot assign","highlight_start":5,"highlight_end":19}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:28:5\n   |\n28 |     || set(&mut x); //~ ERROR cannot assign\n   |     ^^^^^^^^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot assign to immutable item `x`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":708,"byte_end":713,"line_start":19,"line_end":19,"column_start":13,"column_end":18,"is_primary":true,"text":[{"text":"    move || x = 1; //~ ERROR cannot assign","highlight_start":13,"highlight_end":18}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:19:13\n   |\n19 |     move || x = 1; //~ ERROR cannot assign\n   |             ^^^^^ cannot mutate\n\n"}
{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":755,"byte_end":761,"line_start":20,"line_end":20,"column_start":17,"column_end":23,"is_primary":true,"text":[{"text":"    move || set(&mut x); //~ ERROR cannot borrow","highlight_start":17,"highlight_end":23}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:20:17\n   |\n20 |     move || set(&mut x); //~ ERROR cannot borrow\n   |                 ^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot assign to immutable item `x`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":800,"byte_end":805,"line_start":21,"line_end":21,"column_start":13,"column_end":18,"is_primary":true,"text":[{"text":"    move || x = 1; //~ ERROR cannot assign","highlight_start":13,"highlight_end":18}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:21:13\n   |\n21 |     move || x = 1; //~ ERROR cannot assign\n   |             ^^^^^ cannot mutate\n\n"}
{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":847,"byte_end":853,"line_start":22,"line_end":22,"column_start":17,"column_end":23,"is_primary":true,"text":[{"text":"    move || set(&mut x); //~ ERROR cannot borrow","highlight_start":17,"highlight_end":23}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:22:17\n   |\n22 |     move || set(&mut x); //~ ERROR cannot borrow\n   |                 ^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot assign to immutable item `x`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":887,"byte_end":892,"line_start":23,"line_end":23,"column_start":8,"column_end":13,"is_primary":true,"text":[{"text":"    || x = 1; //~ ERROR cannot assign","highlight_start":8,"highlight_end":13}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `x`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:23:8\n   |\n23 |     || x = 1; //~ ERROR cannot assign\n   |        ^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `x`\n\n"}
{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":989,"byte_end":995,"line_start":25,"line_end":25,"column_start":12,"column_end":18,"is_primary":true,"text":[{"text":"    || set(&mut x); //~ ERROR cannot assign","highlight_start":12,"highlight_end":18}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `x`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:25:12\n   |\n25 |     || set(&mut x); //~ ERROR cannot assign\n   |            ^^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `x`\n\n"}
{"message":"cannot assign to immutable item `x`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":1029,"byte_end":1034,"line_start":26,"line_end":26,"column_start":8,"column_end":13,"is_primary":true,"text":[{"text":"    || x = 1; //~ ERROR cannot assign","highlight_start":8,"highlight_end":13}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `x`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:26:8\n   |\n26 |     || x = 1; //~ ERROR cannot assign\n   |        ^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `x`\n\n"}
{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs","byte_start":1131,"byte_end":1137,"line_start":28,"line_end":28,"column_start":12,"column_end":18,"is_primary":true,"text":[{"text":"    || set(&mut x); //~ ERROR cannot assign","highlight_start":12,"highlight_end":18}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `x`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/unboxed-closure-immutable-capture.rs:28:12\n   |\n28 |     || set(&mut x); //~ ERROR cannot assign\n   |            ^^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `x`\n\n"}
{"message":"aborting due to 12 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 12 previous errors\n\n"}
