{"message":"cannot borrow `xs` as mutable more than once at a time (Ast)","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/vec-mut-iter-borrow.rs","byte_start":535,"byte_end":537,"line_start":14,"line_end":14,"column_start":19,"column_end":21,"is_primary":false,"text":[{"text":"    for x in &mut xs {","highlight_start":19,"highlight_end":21}],"label":"first mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/vec-mut-iter-borrow.rs","byte_start":548,"byte_end":550,"line_start":15,"line_end":15,"column_start":9,"column_end":11,"is_primary":true,"text":[{"text":"        xs.push(1) //~ ERROR cannot borrow `xs`","highlight_start":9,"highlight_end":11}],"label":"second mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/vec-mut-iter-borrow.rs","byte_start":592,"byte_end":593,"line_start":16,"line_end":16,"column_start":5,"column_end":6,"is_primary":false,"text":[{"text":"    }","highlight_start":5,"highlight_end":6}],"label":"first borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot assign twice to immutable variable `x` (Mir)","code":{"code":"E0384","explanation":"\nThis error occurs when an attempt is made to reassign an immutable variable.\nFor example:\n\n```compile_fail,E0384\nfn main() {\n    let x = 3;\n    x = 5; // error, reassignment of immutable variable\n}\n```\n\nBy default, variables in Rust are immutable. To fix this error, add the keyword\n`mut` after the keyword `let` when declaring the variable. For example:\n\n```\nfn main() {\n    let mut x = 3;\n    x = 5;\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/vec-mut-iter-borrow.rs","byte_start":521,"byte_end":593,"line_start":14,"line_end":16,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    for x in &mut xs {","highlight_start":5,"highlight_end":23},{"text":"        xs.push(1) //~ ERROR cannot borrow `xs`","highlight_start":1,"highlight_end":48},{"text":"    }","highlight_start":1,"highlight_end":6}],"label":"cannot assign twice to immutable variable","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/vec-mut-iter-borrow.rs","byte_start":521,"byte_end":593,"line_start":14,"line_end":16,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    for x in &mut xs {","highlight_start":5,"highlight_end":23},{"text":"        xs.push(1) //~ ERROR cannot borrow `xs`","highlight_start":1,"highlight_end":48},{"text":"    }","highlight_start":1,"highlight_end":6}],"label":"first assignment to `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `xs` as mutable more than once at a time (Mir)","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/vec-mut-iter-borrow.rs","byte_start":530,"byte_end":537,"line_start":14,"line_end":14,"column_start":14,"column_end":21,"is_primary":false,"text":[{"text":"    for x in &mut xs {","highlight_start":14,"highlight_end":21}],"label":"first mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/vec-mut-iter-borrow.rs","byte_start":548,"byte_end":550,"line_start":15,"line_end":15,"column_start":9,"column_end":11,"is_primary":true,"text":[{"text":"        xs.push(1) //~ ERROR cannot borrow `xs`","highlight_start":9,"highlight_end":11}],"label":"second mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/vec-mut-iter-borrow.rs","byte_start":592,"byte_end":593,"line_start":16,"line_end":16,"column_start":5,"column_end":6,"is_primary":false,"text":[{"text":"    }","highlight_start":5,"highlight_end":6}],"label":"first borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 3 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
