{"message":"cannot borrow immutable local variable `v` as mutable (Ast)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/writing-to-immutable-vec.rs","byte_start":523,"byte_end":524,"line_start":14,"line_end":14,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    v[1] = 4; //~ ERROR cannot borrow immutable local variable `v` as mutable","highlight_start":5,"highlight_end":6}],"label":"cannot borrow mutably","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/writing-to-immutable-vec.rs","byte_start":488,"byte_end":489,"line_start":13,"line_end":13,"column_start":9,"column_end":10,"is_primary":false,"text":[{"text":"    let v: Vec<isize> = vec![1, 2, 3];","highlight_start":9,"highlight_end":10}],"label":"consider changing this to `mut v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow immutable item `v` as mutable (Mir)","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/writing-to-immutable-vec.rs","byte_start":523,"byte_end":524,"line_start":14,"line_end":14,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    v[1] = 4; //~ ERROR cannot borrow immutable local variable `v` as mutable","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
