{"message":"cannot borrow `p` as mutable because it is also borrowed as immutable (Ast)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":935,"byte_end":937,"line_start":34,"line_end":34,"column_start":14,"column_end":16,"is_primary":true,"text":[{"text":"    p.blockm(|| { //~ ERROR cannot borrow `p` as mutable","highlight_start":14,"highlight_end":16}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":926,"byte_end":927,"line_start":34,"line_end":34,"column_start":5,"column_end":6,"is_primary":false,"text":[{"text":"    p.blockm(|| { //~ ERROR cannot borrow `p` as mutable","highlight_start":5,"highlight_end":6}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":1002,"byte_end":1003,"line_start":36,"line_end":36,"column_start":6,"column_end":7,"is_primary":false,"text":[{"text":"    })","highlight_start":6,"highlight_end":7}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":987,"byte_end":988,"line_start":35,"line_end":35,"column_start":9,"column_end":10,"is_primary":false,"text":[{"text":"        p.x = 10;","highlight_start":9,"highlight_end":10}],"label":"borrow occurs due to use of `p` in closure","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `p` as immutable because it is also borrowed as mutable (Ast)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":1152,"byte_end":1153,"line_start":45,"line_end":45,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    p.impurem(); //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":1145,"byte_end":1146,"line_start":44,"line_end":44,"column_start":18,"column_end":19,"is_primary":false,"text":[{"text":"    let l = &mut p;","highlight_start":18,"highlight_end":19}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":1204,"byte_end":1205,"line_start":48,"line_end":48,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"mutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `p` as mutable because it is also borrowed as immutable (Mir)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":935,"byte_end":1002,"line_start":34,"line_end":36,"column_start":14,"column_end":6,"is_primary":true,"text":[{"text":"    p.blockm(|| { //~ ERROR cannot borrow `p` as mutable","highlight_start":14,"highlight_end":57},{"text":"        p.x = 10;","highlight_start":1,"highlight_end":18},{"text":"    })","highlight_start":1,"highlight_end":6}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":926,"byte_end":927,"line_start":34,"line_end":34,"column_start":5,"column_end":6,"is_primary":false,"text":[{"text":"    p.blockm(|| { //~ ERROR cannot borrow `p` as mutable","highlight_start":5,"highlight_end":6}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":1002,"byte_end":1003,"line_start":36,"line_end":36,"column_start":6,"column_end":7,"is_primary":false,"text":[{"text":"    })","highlight_start":6,"highlight_end":7}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `p` as immutable because it is also borrowed as mutable (Mir)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":1152,"byte_end":1153,"line_start":45,"line_end":45,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    p.impurem(); //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":1140,"byte_end":1146,"line_start":44,"line_end":44,"column_start":13,"column_end":19,"is_primary":false,"text":[{"text":"    let l = &mut p;","highlight_start":13,"highlight_end":19}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-rcvr.rs","byte_start":1204,"byte_end":1205,"line_start":48,"line_end":48,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"mutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
