{"message":"cannot assign to immutable item `*s.pointer`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-to-andmut-in-aliasable-loc.rs","byte_start":660,"byte_end":675,"line_start":19,"line_end":19,"column_start":5,"column_end":20,"is_primary":true,"text":[{"text":"    *s.pointer += 1; //~ ERROR cannot assign","highlight_start":5,"highlight_end":20}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*s`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `*s.pointer`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-to-andmut-in-aliasable-loc.rs:19:5\n   |\n19 |     *s.pointer += 1; //~ ERROR cannot assign\n   |     ^^^^^^^^^^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `*s`\n\n"}
{"message":"cannot assign to immutable item `*s.pointer`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-to-andmut-in-aliasable-loc.rs","byte_start":770,"byte_end":785,"line_start":27,"line_end":27,"column_start":5,"column_end":20,"is_primary":true,"text":[{"text":"    *s.pointer += 1; //~ ERROR cannot assign","highlight_start":5,"highlight_end":20}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*s`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `*s.pointer`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-to-andmut-in-aliasable-loc.rs:27:5\n   |\n27 |     *s.pointer += 1; //~ ERROR cannot assign\n   |     ^^^^^^^^^^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `*s`\n\n"}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
