{"message":"cannot assign to immutable item `**t1`","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-mut-base-ptr-in-aliasable-loc.rs","byte_start":699,"byte_end":708,"line_start":19,"line_end":19,"column_start":5,"column_end":14,"is_primary":true,"text":[{"text":"    **t1 = 22; //~ ERROR cannot assign","highlight_start":5,"highlight_end":14}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*t1`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0594]: cannot assign to immutable item `**t1`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-mut-base-ptr-in-aliasable-loc.rs:19:5\n   |\n19 |     **t1 = 22; //~ ERROR cannot assign\n   |     ^^^^^^^^^ cannot mutate\n   |\n   = note: Value not mutable causing this error: `*t1`\n\n"}
{"message":"cannot borrow `**t0` as immutable because it is also borrowed as mutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-mut-base-ptr-in-aliasable-loc.rs","byte_start":811,"byte_end":816,"line_start":24,"line_end":24,"column_start":21,"column_end":26,"is_primary":true,"text":[{"text":"    let p: &isize = &**t0; //~ ERROR cannot borrow","highlight_start":21,"highlight_end":26}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-mut-base-ptr-in-aliasable-loc.rs","byte_start":781,"byte_end":789,"line_start":23,"line_end":23,"column_start":14,"column_end":22,"is_primary":false,"text":[{"text":"    let t1 = &mut *t0;","highlight_start":14,"highlight_end":22}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-mut-base-ptr-in-aliasable-loc.rs","byte_start":857,"byte_end":858,"line_start":26,"line_end":26,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"mutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `**t0` as immutable because it is also borrowed as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-mut-base-ptr-in-aliasable-loc.rs:24:21\n   |\n23 |     let t1 = &mut *t0;\n   |              -------- mutable borrow occurs here\n24 |     let p: &isize = &**t0; //~ ERROR cannot borrow\n   |                     ^^^^^ immutable borrow occurs here\n25 |     **t1 = 22;\n26 | }\n   | - mutable borrow ends here\n\n"}
{"message":"cannot borrow immutable item `**t0` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-mut-base-ptr-in-aliasable-loc.rs","byte_start":913,"byte_end":922,"line_start":29,"line_end":29,"column_start":26,"column_end":35,"is_primary":true,"text":[{"text":"    let x:  &mut isize = &mut **t0; //~ ERROR cannot borrow","highlight_start":26,"highlight_end":35}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*t0`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `**t0` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-mut-base-ptr-in-aliasable-loc.rs:29:26\n   |\n29 |     let x:  &mut isize = &mut **t0; //~ ERROR cannot borrow\n   |                          ^^^^^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `*t0`\n\n"}
{"message":"aborting due to 3 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 3 previous errors\n\n"}
