{"message":"borrow of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-block-unint.rs","byte_start":552,"byte_end":648,"line_start":14,"line_end":16,"column_start":11,"column_end":6,"is_primary":true,"text":[{"text":"    force(|| {  //~ ERROR capture of possibly uninitialized variable: `x`","highlight_start":11,"highlight_end":74},{"text":"        println!(\"{}\", x);","highlight_start":1,"highlight_end":27},{"text":"    });","highlight_start":1,"highlight_end":6}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: borrow of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-block-unint.rs:14:11\n   |\n14 |       force(|| {  //~ ERROR capture of possibly uninitialized variable: `x`\n   |  ___________^\n15 | |         println!(\"{}\", x);\n16 | |     });\n   | |_____^ use of possibly uninitialized `x`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
