{"message":"use of possibly uninitialized variable: `u.a`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-union-uninitialized.rs","byte_start":714,"byte_end":717,"line_start":26,"line_end":26,"column_start":18,"column_end":21,"is_primary":true,"text":[{"text":"        let ua = u.a; //~ ERROR use of possibly uninitialized variable: `u.a`","highlight_start":18,"highlight_end":21}],"label":"use of possibly uninitialized `u.a`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `u.a`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-union-uninitialized.rs:26:18\n   |\n26 |         let ua = u.a; //~ ERROR use of possibly uninitialized variable: `u.a`\n   |                  ^^^ use of possibly uninitialized `u.a`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
