{"message":"use of possibly uninitialized variable: `v` (Ast)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-while-break.rs","byte_start":572,"byte_end":573,"line_start":17,"line_end":17,"column_start":20,"column_end":21,"is_primary":true,"text":[{"text":"    println!(\"{}\", v); //~ ERROR use of possibly uninitialized variable: `v`","highlight_start":20,"highlight_end":21}],"label":"use of possibly uninitialized `v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"borrow of possibly uninitialized variable: `v` (Mir)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-emergency/src/test/compile-fail/borrowck/borrowck-while-break.rs","byte_start":572,"byte_end":573,"line_start":17,"line_end":17,"column_start":20,"column_end":21,"is_primary":true,"text":[{"text":"    println!(\"{}\", v); //~ ERROR use of possibly uninitialized variable: `v`","highlight_start":20,"highlight_end":21}],"label":"use of possibly uninitialized `v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
