{"message":"cannot borrow `*v` as mutable because it is also borrowed as immutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-if.rs","byte_start":1236,"byte_end":1243,"line_start":38,"line_end":38,"column_start":16,"column_end":23,"is_primary":true,"text":[{"text":"    borrow_mut(&mut *v); //~ ERROR cannot borrow","highlight_start":16,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-if.rs","byte_start":1206,"byte_end":1213,"line_start":36,"line_end":36,"column_start":9,"column_end":16,"is_primary":false,"text":[{"text":"        _w = &v;","highlight_start":9,"highlight_end":16}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-if.rs","byte_start":1270,"byte_end":1271,"line_start":39,"line_end":39,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `*v` as mutable because it is also borrowed as immutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-if.rs:38:16\n   |\n36 |         _w = &v;\n   |         ------- immutable borrow occurs here\n37 |     }\n38 |     borrow_mut(&mut *v); //~ ERROR cannot borrow\n   |                ^^^^^^^ mutable borrow occurs here\n39 | }\n   | - immutable borrow ends here\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
