{"message":"cannot borrow `a` as mutable because `*a` is also borrowed as immutable (Ast)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":988,"byte_end":989,"line_start":35,"line_end":35,"column_start":22,"column_end":23,"is_primary":true,"text":[{"text":"        rewrite(&mut a)); //~ ERROR cannot borrow","highlight_start":22,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":963,"byte_end":965,"line_start":34,"line_end":34,"column_start":10,"column_end":12,"is_primary":false,"text":[{"text":"        &*a,","highlight_start":10,"highlight_end":12}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":990,"byte_end":991,"line_start":35,"line_end":35,"column_start":24,"column_end":25,"is_primary":false,"text":[{"text":"        rewrite(&mut a)); //~ ERROR cannot borrow","highlight_start":24,"highlight_end":25}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `a` as mutable because `*a` is also borrowed as immutable (Ast)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":1110,"byte_end":1111,"line_start":42,"line_end":42,"column_start":22,"column_end":23,"is_primary":true,"text":[{"text":"        rewrite(&mut a)); //~ ERROR cannot borrow","highlight_start":22,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":1085,"byte_end":1087,"line_start":41,"line_end":41,"column_start":10,"column_end":12,"is_primary":false,"text":[{"text":"        &*a,","highlight_start":10,"highlight_end":12}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":1112,"byte_end":1113,"line_start":42,"line_end":42,"column_start":24,"column_end":25,"is_primary":false,"text":[{"text":"        rewrite(&mut a)); //~ ERROR cannot borrow","highlight_start":24,"highlight_end":25}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `a` as mutable because it is also borrowed as immutable (Mir)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":983,"byte_end":989,"line_start":35,"line_end":35,"column_start":17,"column_end":23,"is_primary":true,"text":[{"text":"        rewrite(&mut a)); //~ ERROR cannot borrow","highlight_start":17,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":962,"byte_end":965,"line_start":34,"line_end":34,"column_start":9,"column_end":12,"is_primary":false,"text":[{"text":"        &*a,","highlight_start":9,"highlight_end":12}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":990,"byte_end":991,"line_start":35,"line_end":35,"column_start":24,"column_end":25,"is_primary":false,"text":[{"text":"        rewrite(&mut a)); //~ ERROR cannot borrow","highlight_start":24,"highlight_end":25}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `a` as mutable because it is also borrowed as immutable (Mir)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":1105,"byte_end":1111,"line_start":42,"line_end":42,"column_start":17,"column_end":23,"is_primary":true,"text":[{"text":"        rewrite(&mut a)); //~ ERROR cannot borrow","highlight_start":17,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":1084,"byte_end":1087,"line_start":41,"line_end":41,"column_start":9,"column_end":12,"is_primary":false,"text":[{"text":"        &*a,","highlight_start":9,"highlight_end":12}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-bad-nested-calls-free.rs","byte_start":1112,"byte_end":1113,"line_start":42,"line_end":42,"column_start":24,"column_end":25,"is_primary":false,"text":[{"text":"        rewrite(&mut a)); //~ ERROR cannot borrow","highlight_start":24,"highlight_end":25}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
