{"message":"use of possibly uninitialized variable: `bar` (Ast)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-after-item.rs","byte_start":526,"byte_end":529,"line_start":14,"line_end":14,"column_start":9,"column_end":12,"is_primary":true,"text":[{"text":"    baz(bar); //~ ERROR use of possibly uninitialized variable: `bar`","highlight_start":9,"highlight_end":12}],"label":"use of possibly uninitialized `bar`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"use of possibly uninitialized variable: `bar` (Mir)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-after-item.rs","byte_start":526,"byte_end":529,"line_start":14,"line_end":14,"column_start":9,"column_end":12,"is_primary":true,"text":[{"text":"    baz(bar); //~ ERROR use of possibly uninitialized variable: `bar`","highlight_start":9,"highlight_end":12}],"label":"use of possibly uninitialized `bar`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
