{"message":"cannot borrow `*v` as immutable because it is also borrowed as mutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1157,"byte_end":1160,"line_start":35,"line_end":35,"column_start":16,"column_end":19,"is_primary":true,"text":[{"text":"        borrow(&*v); //~ ERROR cannot borrow","highlight_start":16,"highlight_end":19}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1109,"byte_end":1115,"line_start":32,"line_end":32,"column_start":17,"column_end":23,"is_primary":false,"text":[{"text":"    let mut x = &mut v;","highlight_start":17,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1193,"byte_end":1194,"line_start":37,"line_end":37,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"mutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `*v` as immutable because it is also borrowed as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs:35:16\n   |\n32 |     let mut x = &mut v;\n   |                 ------ mutable borrow occurs here\n...\n35 |         borrow(&*v); //~ ERROR cannot borrow\n   |                ^^^ immutable borrow occurs here\n36 |     }\n37 | }\n   | - mutable borrow ends here\n\n"}
{"message":"cannot borrow `*v` as immutable because it is also borrowed as mutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1395,"byte_end":1398,"line_start":45,"line_end":45,"column_start":16,"column_end":19,"is_primary":true,"text":[{"text":"        borrow(&*v); //~ ERROR cannot borrow","highlight_start":16,"highlight_end":19}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1352,"byte_end":1358,"line_start":43,"line_end":43,"column_start":17,"column_end":23,"is_primary":false,"text":[{"text":"    let mut x = &mut v;","highlight_start":17,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1447,"byte_end":1448,"line_start":48,"line_end":48,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"mutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `*v` as immutable because it is also borrowed as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs:45:16\n   |\n43 |     let mut x = &mut v;\n   |                 ------ mutable borrow occurs here\n44 |     for _ in 0..3 {\n45 |         borrow(&*v); //~ ERROR cannot borrow\n   |                ^^^ immutable borrow occurs here\n...\n48 | }\n   | - mutable borrow ends here\n\n"}
{"message":"cannot borrow `*v` as mutable because it is also borrowed as immutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1653,"byte_end":1660,"line_start":57,"line_end":57,"column_start":20,"column_end":27,"is_primary":true,"text":[{"text":"        borrow_mut(&mut *v); //~ ERROR cannot borrow","highlight_start":20,"highlight_end":27}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1700,"byte_end":1702,"line_start":58,"line_end":58,"column_start":14,"column_end":16,"is_primary":false,"text":[{"text":"        _x = &v;","highlight_start":14,"highlight_end":16}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1710,"byte_end":1711,"line_start":60,"line_end":60,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `*v` as mutable because it is also borrowed as immutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs:57:20\n   |\n57 |         borrow_mut(&mut *v); //~ ERROR cannot borrow\n   |                    ^^^^^^^ mutable borrow occurs here\n58 |         _x = &v;\n   |              -- immutable borrow occurs here\n59 |     }\n60 | }\n   | - immutable borrow ends here\n\n"}
{"message":"cannot borrow `*v` as mutable because it is also borrowed as immutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1925,"byte_end":1932,"line_start":69,"line_end":69,"column_start":20,"column_end":27,"is_primary":true,"text":[{"text":"        borrow_mut(&mut *v); //~ ERROR cannot borrow","highlight_start":20,"highlight_end":27}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1972,"byte_end":1974,"line_start":70,"line_end":70,"column_start":14,"column_end":16,"is_primary":false,"text":[{"text":"        _x = &v;","highlight_start":14,"highlight_end":16}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":1982,"byte_end":1983,"line_start":72,"line_end":72,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `*v` as mutable because it is also borrowed as immutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs:69:20\n   |\n69 |         borrow_mut(&mut *v); //~ ERROR cannot borrow\n   |                    ^^^^^^^ mutable borrow occurs here\n70 |         _x = &v;\n   |              -- immutable borrow occurs here\n71 |     }\n72 | }\n   | - immutable borrow ends here\n\n"}
{"message":"cannot borrow `*v` as mutable because it is also borrowed as immutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2258,"byte_end":2265,"line_start":86,"line_end":86,"column_start":16,"column_end":23,"is_primary":true,"text":[{"text":"    borrow_mut(&mut *v); //~ ERROR cannot borrow","highlight_start":16,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2218,"byte_end":2220,"line_start":83,"line_end":83,"column_start":14,"column_end":16,"is_primary":false,"text":[{"text":"        _x = &v;","highlight_start":14,"highlight_end":16}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2292,"byte_end":2293,"line_start":87,"line_end":87,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `*v` as mutable because it is also borrowed as immutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs:86:16\n   |\n83 |         _x = &v;\n   |              -- immutable borrow occurs here\n...\n86 |     borrow_mut(&mut *v); //~ ERROR cannot borrow\n   |                ^^^^^^^ mutable borrow occurs here\n87 | }\n   | - immutable borrow ends here\n\n"}
{"message":"cannot borrow `*v` as mutable because it is also borrowed as immutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2576,"byte_end":2583,"line_start":100,"line_end":100,"column_start":16,"column_end":23,"is_primary":true,"text":[{"text":"    borrow_mut(&mut *v); //~ ERROR cannot borrow","highlight_start":16,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2536,"byte_end":2538,"line_start":97,"line_end":97,"column_start":14,"column_end":16,"is_primary":false,"text":[{"text":"        _x = &v;","highlight_start":14,"highlight_end":16}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2610,"byte_end":2611,"line_start":101,"line_end":101,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `*v` as mutable because it is also borrowed as immutable\n   --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs:100:16\n    |\n97  |         _x = &v;\n    |              -- immutable borrow occurs here\n...\n100 |     borrow_mut(&mut *v); //~ ERROR cannot borrow\n    |                ^^^^^^^ mutable borrow occurs here\n101 | }\n    | - immutable borrow ends here\n\n"}
{"message":"cannot borrow `*v` as immutable because it is also borrowed as mutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2802,"byte_end":2805,"line_start":109,"line_end":109,"column_start":16,"column_end":19,"is_primary":true,"text":[{"text":"        borrow(&*v); //~ ERROR cannot borrow","highlight_start":16,"highlight_end":19}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2867,"byte_end":2873,"line_start":111,"line_end":111,"column_start":17,"column_end":23,"is_primary":false,"text":[{"text":"            x = &mut v; //~ ERROR cannot borrow","highlight_start":17,"highlight_end":23}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2915,"byte_end":2916,"line_start":114,"line_end":114,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"mutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `*v` as immutable because it is also borrowed as mutable\n   --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs:109:16\n    |\n109 |         borrow(&*v); //~ ERROR cannot borrow\n    |                ^^^ immutable borrow occurs here\n110 |         if cond2 {\n111 |             x = &mut v; //~ ERROR cannot borrow\n    |                 ------ mutable borrow occurs here\n...\n114 | }\n    | - mutable borrow ends here\n\n"}
{"message":"cannot borrow `v` as mutable more than once at a time","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2867,"byte_end":2873,"line_start":111,"line_end":111,"column_start":17,"column_end":23,"is_primary":true,"text":[{"text":"            x = &mut v; //~ ERROR cannot borrow","highlight_start":17,"highlight_end":23}],"label":"mutable borrow starts here in previous iteration of loop","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs","byte_start":2915,"byte_end":2916,"line_start":114,"line_end":114,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"mutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0499]: cannot borrow `v` as mutable more than once at a time\n   --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-lend-flow-loop.rs:111:17\n    |\n111 |             x = &mut v; //~ ERROR cannot borrow\n    |                 ^^^^^^ mutable borrow starts here in previous iteration of loop\n...\n114 | }\n    | - mutable borrow ends here\n\n"}
{"message":"aborting due to 8 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 8 previous errors\n\n"}
