{"message":"borrowed value does not live long enough","code":{"code":"E0597","explanation":"\nThis error occurs because a borrow was made inside a variable which has a\ngreater lifetime than the borrowed one.\n\nExample of erroneous code:\n\n```compile_fail,E0597\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet mut x = Foo { x: None };\nlet y = 0;\nx.x = Some(&y); // error: `y` does not live long enough\n```\n\nIn here, `x` is created before `y` and therefore has a greater lifetime. Always\nkeep in mind that values in a scope are dropped in the opposite order they are\ncreated. So to fix the previous example, just make the `y` lifetime greater than\nthe `x`'s one:\n\n```\nstruct Foo<'a> {\n    x: Option<&'a u32>,\n}\n\nlet y = 0;\nlet mut x = Foo { x: None };\nx.x = Some(&y);\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrowed-uniq-rvalue.rs","byte_start":660,"byte_end":671,"line_start":20,"line_end":20,"column_start":28,"column_end":39,"is_primary":false,"text":[{"text":"    buggy_map.insert(42, &*Box::new(1)); //~ ERROR borrowed value does not live long enough","highlight_start":28,"highlight_end":39}],"label":"temporary value created here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrowed-uniq-rvalue.rs","byte_start":673,"byte_end":673,"line_start":20,"line_end":20,"column_start":41,"column_end":41,"is_primary":true,"text":[{"text":"    buggy_map.insert(42, &*Box::new(1)); //~ ERROR borrowed value does not live long enough","highlight_start":41,"highlight_end":41}],"label":"temporary value dropped here while still borrowed","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrowed-uniq-rvalue.rs","byte_start":818,"byte_end":819,"line_start":25,"line_end":25,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"temporary value needs to live until here","suggested_replacement":null,"expansion":null}],"children":[{"message":"consider using a `let` binding to increase its lifetime","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0597]: borrowed value does not live long enough\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrowed-uniq-rvalue.rs:20:41\n   |\n20 |     buggy_map.insert(42, &*Box::new(1)); //~ ERROR borrowed value does not live long enough\n   |                            -----------  ^ temporary value dropped here while still borrowed\n   |                            |\n   |                            temporary value created here\n...\n25 | }\n   | - temporary value needs to live until here\n   |\n   = note: consider using a `let` binding to increase its lifetime\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
