{"message":"cannot borrow `vector` as mutable because it is also borrowed as immutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-for-loop-head-linkage.rs","byte_start":605,"byte_end":611,"line_start":17,"line_end":17,"column_start":9,"column_end":15,"is_primary":true,"text":[{"text":"        vector.extend(repeat(0));      //~ ERROR cannot borrow","highlight_start":9,"highlight_end":15}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-for-loop-head-linkage.rs","byte_start":550,"byte_end":557,"line_start":15,"line_end":15,"column_start":15,"column_end":22,"is_primary":false,"text":[{"text":"    for &x in &vector {","highlight_start":15,"highlight_end":22}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-for-loop-head-linkage.rs","byte_start":713,"byte_end":714,"line_start":19,"line_end":19,"column_start":5,"column_end":6,"is_primary":false,"text":[{"text":"    }","highlight_start":5,"highlight_end":6}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `vector` as mutable because it is also borrowed as immutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-for-loop-head-linkage.rs:17:9\n   |\n15 |     for &x in &vector {\n   |               ------- immutable borrow occurs here\n16 |         let cap = vector.capacity();\n17 |         vector.extend(repeat(0));      //~ ERROR cannot borrow\n   |         ^^^^^^ mutable borrow occurs here\n18 |         vector[1] = 5;   //~ ERROR cannot borrow\n19 |     }\n   |     - immutable borrow ends here\n\n"}
{"message":"cannot borrow `vector` as mutable because it is also borrowed as immutable","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-for-loop-head-linkage.rs","byte_start":668,"byte_end":674,"line_start":18,"line_end":18,"column_start":9,"column_end":15,"is_primary":true,"text":[{"text":"        vector[1] = 5;   //~ ERROR cannot borrow","highlight_start":9,"highlight_end":15}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-for-loop-head-linkage.rs","byte_start":550,"byte_end":557,"line_start":15,"line_end":15,"column_start":15,"column_end":22,"is_primary":false,"text":[{"text":"    for &x in &vector {","highlight_start":15,"highlight_end":22}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-for-loop-head-linkage.rs","byte_start":713,"byte_end":714,"line_start":19,"line_end":19,"column_start":5,"column_end":6,"is_primary":false,"text":[{"text":"    }","highlight_start":5,"highlight_end":6}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0502]: cannot borrow `vector` as mutable because it is also borrowed as immutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-for-loop-head-linkage.rs:18:9\n   |\n15 |     for &x in &vector {\n   |               ------- immutable borrow occurs here\n...\n18 |         vector[1] = 5;   //~ ERROR cannot borrow\n   |         ^^^^^^ mutable borrow occurs here\n19 |     }\n   |     - immutable borrow ends here\n\n"}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
