{"message":"cannot borrow `x` as mutable more than once at a time (Ast)","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":764,"byte_end":765,"line_start":24,"line_end":24,"column_start":11,"column_end":12,"is_primary":false,"text":[{"text":"    (&mut x).with(","highlight_start":11,"highlight_end":12}],"label":"first mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":781,"byte_end":786,"line_start":25,"line_end":25,"column_start":9,"column_end":14,"is_primary":true,"text":[{"text":"        |opt| { //~ ERROR cannot borrow `x` as mutable more than once at a time","highlight_start":9,"highlight_end":14}],"label":"second mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":1068,"byte_end":1069,"line_start":33,"line_end":33,"column_start":10,"column_end":11,"is_primary":false,"text":[{"text":"        })","highlight_start":10,"highlight_end":11}],"label":"first borrow ends here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":940,"byte_end":941,"line_start":28,"line_end":28,"column_start":21,"column_end":22,"is_primary":false,"text":[{"text":"                    x = X(Either::Left((0, 0)));","highlight_start":21,"highlight_end":22}],"label":"borrow occurs due to use of `x` in closure","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `x` as mutable more than once at a time (Mir)","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":758,"byte_end":766,"line_start":24,"line_end":24,"column_start":5,"column_end":13,"is_primary":false,"text":[{"text":"    (&mut x).with(","highlight_start":5,"highlight_end":13}],"label":"first mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":781,"byte_end":1068,"line_start":25,"line_end":33,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        |opt| { //~ ERROR cannot borrow `x` as mutable more than once at a time","highlight_start":9,"highlight_end":80},{"text":"            match opt {","highlight_start":1,"highlight_end":24},{"text":"                &Either::Right(ref f) => {","highlight_start":1,"highlight_end":43},{"text":"                    x = X(Either::Left((0, 0)));","highlight_start":1,"highlight_end":49},{"text":"                    (*f)()","highlight_start":1,"highlight_end":27},{"text":"                },","highlight_start":1,"highlight_end":19},{"text":"                _ => panic!()","highlight_start":1,"highlight_end":30},{"text":"            }","highlight_start":1,"highlight_end":14},{"text":"        })","highlight_start":1,"highlight_end":10}],"label":"second mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":1068,"byte_end":1069,"line_start":33,"line_end":33,"column_start":10,"column_end":11,"is_primary":false,"text":[{"text":"        })","highlight_start":10,"highlight_end":11}],"label":"first borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
