{"message":"borrow of possibly uninitialized variable: `v`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-while-break.rs","byte_start":572,"byte_end":573,"line_start":17,"line_end":17,"column_start":20,"column_end":21,"is_primary":true,"text":[{"text":"    println!(\"{}\", v); //~ ERROR use of possibly uninitialized variable: `v`","highlight_start":20,"highlight_end":21}],"label":"use of possibly uninitialized `v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: borrow of possibly uninitialized variable: `v`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-while-break.rs:17:20\n   |\n17 |     println!(\"{}\", v); //~ ERROR use of possibly uninitialized variable: `v`\n   |                    ^ use of possibly uninitialized `v`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
