{"message":"cannot borrow `t0` as mutable because `*t0` is also borrowed as immutable (Ast)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":767,"byte_end":769,"line_start":19,"line_end":19,"column_start":23,"column_end":25,"is_primary":true,"text":[{"text":"    let mut t2 = &mut t0;   //~ ERROR cannot borrow `t0`","highlight_start":23,"highlight_end":25}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":719,"byte_end":722,"line_start":18,"line_end":18,"column_start":22,"column_end":25,"is_primary":false,"text":[{"text":"    let p: &isize = &*t0;     // Freezes `*t0`","highlight_start":22,"highlight_end":25}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":847,"byte_end":848,"line_start":21,"line_end":21,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `t0` as mutable more than once at a time (Ast)","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":949,"byte_end":952,"line_start":25,"line_end":25,"column_start":30,"column_end":33,"is_primary":false,"text":[{"text":"    let p: &mut isize = &mut *t0; // Claims `*t0`","highlight_start":30,"highlight_end":33}],"label":"first mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":992,"byte_end":994,"line_start":26,"line_end":26,"column_start":23,"column_end":25,"is_primary":true,"text":[{"text":"    let mut t2 = &mut t0;       //~ ERROR cannot borrow `t0`","highlight_start":23,"highlight_end":25}],"label":"second mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":1101,"byte_end":1102,"line_start":28,"line_end":28,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"first borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `t0` as mutable because it is also borrowed as immutable (Mir)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":762,"byte_end":769,"line_start":19,"line_end":19,"column_start":18,"column_end":25,"is_primary":true,"text":[{"text":"    let mut t2 = &mut t0;   //~ ERROR cannot borrow `t0`","highlight_start":18,"highlight_end":25}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":718,"byte_end":722,"line_start":18,"line_end":18,"column_start":21,"column_end":25,"is_primary":false,"text":[{"text":"    let p: &isize = &*t0;     // Freezes `*t0`","highlight_start":21,"highlight_end":25}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":847,"byte_end":848,"line_start":21,"line_end":21,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `t0` as mutable more than once at a time (Mir)","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":944,"byte_end":952,"line_start":25,"line_end":25,"column_start":25,"column_end":33,"is_primary":false,"text":[{"text":"    let p: &mut isize = &mut *t0; // Claims `*t0`","highlight_start":25,"highlight_end":33}],"label":"first mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":987,"byte_end":994,"line_start":26,"line_end":26,"column_start":18,"column_end":25,"is_primary":true,"text":[{"text":"    let mut t2 = &mut t0;       //~ ERROR cannot borrow `t0`","highlight_start":18,"highlight_end":25}],"label":"second mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs","byte_start":1101,"byte_end":1102,"line_start":28,"line_end":28,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"first borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
