{"message":"use of possibly uninitialized variable: `t`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-move-moved-value-into-closure.rs","byte_start":633,"byte_end":650,"line_start":21,"line_end":21,"column_start":12,"column_end":29,"is_primary":true,"text":[{"text":"    call_f(move|| { *t + 1 }); //~ ERROR capture of moved value","highlight_start":12,"highlight_end":29}],"label":"use of possibly uninitialized `t`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `t`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-move-moved-value-into-closure.rs:21:12\n   |\n21 |     call_f(move|| { *t + 1 }); //~ ERROR capture of moved value\n   |            ^^^^^^^^^^^^^^^^^ use of possibly uninitialized `t`\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
