{"message":"cannot borrow immutable item `arg` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":568,"byte_end":571,"line_start":20,"line_end":20,"column_start":5,"column_end":8,"is_primary":true,"text":[{"text":"    arg.mutate(); //~ ERROR: cannot borrow immutable argument","highlight_start":5,"highlight_end":8}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `arg` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs:20:5\n   |\n20 |     arg.mutate(); //~ ERROR: cannot borrow immutable argument\n   |     ^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `arg` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":677,"byte_end":680,"line_start":25,"line_end":25,"column_start":9,"column_end":12,"is_primary":true,"text":[{"text":"        arg.mutate(); //~ ERROR: cannot borrow immutable argument","highlight_start":9,"highlight_end":12}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `arg` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs:25:9\n   |\n25 |         arg.mutate(); //~ ERROR: cannot borrow immutable argument\n   |         ^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `arg` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":794,"byte_end":797,"line_start":31,"line_end":31,"column_start":9,"column_end":12,"is_primary":true,"text":[{"text":"        arg.mutate(); //~ ERROR: cannot borrow immutable argument","highlight_start":9,"highlight_end":12}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `arg` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs:31:9\n   |\n31 |         arg.mutate(); //~ ERROR: cannot borrow immutable argument\n   |         ^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `arg` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs","byte_start":969,"byte_end":972,"line_start":42,"line_end":42,"column_start":17,"column_end":20,"is_primary":true,"text":[{"text":"    (|arg: S| { arg.mutate() })(s); //~ ERROR: cannot borrow immutable argument","highlight_start":17,"highlight_end":20}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `arg` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-argument.rs:42:17\n   |\n42 |     (|arg: S| { arg.mutate() })(s); //~ ERROR: cannot borrow immutable argument\n   |                 ^^^ cannot borrow as mutable\n\n"}
{"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 4 previous errors\n\n"}
