{"message":"cannot borrow immutable item `_.y` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1208,"byte_end":1216,"line_start":57,"line_end":57,"column_start":19,"column_end":27,"is_primary":true,"text":[{"text":"    let __isize = &mut x.y; //~ ERROR cannot borrow","highlight_start":19,"highlight_end":27}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*_`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `_.y` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:57:19\n   |\n57 |     let __isize = &mut x.y; //~ ERROR cannot borrow\n   |                   ^^^^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `*_`\n\n"}
{"message":"cannot borrow immutable item `_.y` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1303,"byte_end":1311,"line_start":61,"line_end":61,"column_start":19,"column_end":27,"is_primary":true,"text":[{"text":"    let __isize = &mut x.y; //~ ERROR cannot borrow","highlight_start":19,"highlight_end":27}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*_`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `_.y` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:61:19\n   |\n61 |     let __isize = &mut x.y; //~ ERROR cannot borrow\n   |                   ^^^^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `*_`\n\n"}
{"message":"cannot borrow immutable item `_.y` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1463,"byte_end":1471,"line_start":69,"line_end":69,"column_start":5,"column_end":13,"is_primary":true,"text":[{"text":"    &mut x.y //~ ERROR cannot borrow","highlight_start":5,"highlight_end":13}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*_`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `_.y` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:69:5\n   |\n69 |     &mut x.y //~ ERROR cannot borrow\n   |     ^^^^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `*_`\n\n"}
{"message":"cannot borrow immutable item `_.y` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1565,"byte_end":1573,"line_start":73,"line_end":73,"column_start":5,"column_end":13,"is_primary":true,"text":[{"text":"    &mut x.y //~ ERROR cannot borrow","highlight_start":5,"highlight_end":13}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[{"message":"Value not mutable causing this error: `*_`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `_.y` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:73:5\n   |\n73 |     &mut x.y //~ ERROR cannot borrow\n   |     ^^^^^^^^ cannot borrow as mutable\n   |\n   = note: Value not mutable causing this error: `*_`\n\n"}
{"message":"cannot borrow immutable item `*_` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":1951,"byte_end":1952,"line_start":93,"line_end":93,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    x.set(0, 0); //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `*_` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:93:5\n   |\n93 |     x.set(0, 0); //~ ERROR cannot borrow\n   |     ^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `*_` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2036,"byte_end":2037,"line_start":97,"line_end":97,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    x.set(0, 0); //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `*_` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:97:5\n   |\n97 |     x.set(0, 0); //~ ERROR cannot borrow\n   |     ^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `*_` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2206,"byte_end":2207,"line_start":105,"line_end":105,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    x.y_mut() //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `*_` as mutable\n   --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:105:5\n    |\n105 |     x.y_mut() //~ ERROR cannot borrow\n    |     ^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `*_` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2310,"byte_end":2311,"line_start":109,"line_end":109,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    x.y_mut() //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `*_` as mutable\n   --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:109:5\n    |\n109 |     x.y_mut() //~ ERROR cannot borrow\n    |     ^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `*_` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2390,"byte_end":2391,"line_start":113,"line_end":113,"column_start":6,"column_end":7,"is_primary":true,"text":[{"text":"    *x.y_mut() = 3; //~ ERROR cannot borrow","highlight_start":6,"highlight_end":7}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `*_` as mutable\n   --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:113:6\n    |\n113 |     *x.y_mut() = 3; //~ ERROR cannot borrow\n    |      ^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `*_` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2479,"byte_end":2480,"line_start":117,"line_end":117,"column_start":6,"column_end":7,"is_primary":true,"text":[{"text":"    *x.y_mut() = 3; //~ ERROR cannot borrow","highlight_start":6,"highlight_end":7}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `*_` as mutable\n   --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:117:6\n    |\n117 |     *x.y_mut() = 3; //~ ERROR cannot borrow\n    |      ^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item `*_` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs","byte_start":2572,"byte_end":2573,"line_start":121,"line_end":121,"column_start":6,"column_end":7,"is_primary":true,"text":[{"text":"    *x.y_mut() = 3; //~ ERROR cannot borrow","highlight_start":6,"highlight_end":7}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `*_` as mutable\n   --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-auto-deref.rs:121:6\n    |\n121 |     *x.y_mut() = 3; //~ ERROR cannot borrow\n    |      ^ cannot borrow as mutable\n\n"}
{"message":"aborting due to 11 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 11 previous errors\n\n"}
