{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":609,"byte_end":615,"line_start":16,"line_end":16,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    x += 1; //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":11}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:16:5\n   |\n16 |     x += 1; //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":694,"byte_end":700,"line_start":19,"line_end":19,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    x -= 1; //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":11}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:19:5\n   |\n19 |     x -= 1; //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":779,"byte_end":785,"line_start":22,"line_end":22,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    x *= 1; //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":11}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:22:5\n   |\n22 |     x *= 1; //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":864,"byte_end":870,"line_start":25,"line_end":25,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    x /= 1; //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":11}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:25:5\n   |\n25 |     x /= 1; //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":949,"byte_end":955,"line_start":28,"line_end":28,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    x %= 1; //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":11}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:28:5\n   |\n28 |     x %= 1; //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":1034,"byte_end":1040,"line_start":31,"line_end":31,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    x ^= 1; //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":11}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:31:5\n   |\n31 |     x ^= 1; //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":1119,"byte_end":1125,"line_start":34,"line_end":34,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    x &= 1; //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":11}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:34:5\n   |\n34 |     x &= 1; //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":1204,"byte_end":1210,"line_start":37,"line_end":37,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    x |= 1; //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":11}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:37:5\n   |\n37 |     x |= 1; //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":1289,"byte_end":1296,"line_start":40,"line_end":40,"column_start":5,"column_end":12,"is_primary":true,"text":[{"text":"    x <<= 1;    //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":12}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:40:5\n   |\n40 |     x <<= 1;    //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"use of possibly uninitialized variable: `x`","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs","byte_start":1378,"byte_end":1385,"line_start":43,"line_end":43,"column_start":5,"column_end":12,"is_primary":true,"text":[{"text":"    x >>= 1;    //~ ERROR use of possibly uninitialized variable: `x`","highlight_start":5,"highlight_end":12}],"label":"use of possibly uninitialized `x`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0381]: use of possibly uninitialized variable: `x`\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-uninit-in-assignop.rs:43:5\n   |\n43 |     x >>= 1;    //~ ERROR use of possibly uninitialized variable: `x`\n   |     ^^^^^^^ use of possibly uninitialized `x`\n\n"}
{"message":"aborting due to 12 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 12 previous errors\n\n"}
