{"message":"use of possibly uninitialized variable: `v` (Ast)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-init-plus-equal.rs","byte_start":509,"byte_end":510,"line_start":13,"line_end":13,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"    v = v + 1; //~ ERROR use of possibly uninitialized variable: `v`","highlight_start":9,"highlight_end":10}],"label":"use of possibly uninitialized `v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"use of possibly uninitialized variable: `v` (Mir)","code":{"code":"E0381","explanation":"\nIt is not allowed to use or capture an uninitialized variable. For example:\n\n```compile_fail,E0381\nfn main() {\n    let x: i32;\n    let y = x; // error, use of possibly uninitialized variable\n}\n```\n\nTo fix this, ensure that any declared variables are initialized before being\nused. Example:\n\n```\nfn main() {\n    let x: i32 = 0;\n    let y = x; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-init-plus-equal.rs","byte_start":509,"byte_end":510,"line_start":13,"line_end":13,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"    v = v + 1; //~ ERROR use of possibly uninitialized variable: `v`","highlight_start":9,"highlight_end":10}],"label":"use of possibly uninitialized `v`","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
