{"message":"cannot borrow `x` as mutable more than once at a time","code":{"code":"E0499","explanation":"\nA variable was borrowed as mutable more than once. Erroneous code example:\n\n```compile_fail,E0499\nlet mut i = 0;\nlet mut x = &mut i;\nlet mut a = &mut i;\n// error: cannot borrow `i` as mutable more than once at a time\n```\n\nPlease note that in rust, you can either have many immutable references, or one\nmutable reference. Take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html for more\ninformation. Example:\n\n\n```\nlet mut i = 0;\nlet mut x = &mut i; // ok!\n\n// or:\nlet mut i = 0;\nlet a = &i; // ok!\nlet b = &i; // still ok!\nlet c = &i; // still ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":758,"byte_end":766,"line_start":24,"line_end":24,"column_start":5,"column_end":13,"is_primary":false,"text":[{"text":"    (&mut x).with(","highlight_start":5,"highlight_end":13}],"label":"first mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":781,"byte_end":786,"line_start":25,"line_end":25,"column_start":9,"column_end":14,"is_primary":true,"text":[{"text":"        |opt| { //~ ERROR cannot borrow `x` as mutable more than once at a time","highlight_start":9,"highlight_end":14}],"label":"second mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":1068,"byte_end":1069,"line_start":33,"line_end":33,"column_start":10,"column_end":11,"is_primary":false,"text":[{"text":"        })","highlight_start":10,"highlight_end":11}],"label":"first borrow ends here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs","byte_start":940,"byte_end":941,"line_start":28,"line_end":28,"column_start":21,"column_end":22,"is_primary":false,"text":[{"text":"                    x = X(Either::Left((0, 0)));","highlight_start":21,"highlight_end":22}],"label":"borrow occurs due to use of `x` in closure","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0499]: cannot borrow `x` as mutable more than once at a time\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-autoref-3261.rs:25:9\n   |\n24 |     (&mut x).with(\n   |     -------- first mutable borrow occurs here\n25 |         |opt| { //~ ERROR cannot borrow `x` as mutable more than once at a time\n   |         ^^^^^ second mutable borrow occurs here\n...\n28 |                     x = X(Either::Left((0, 0)));\n   |                     - borrow occurs due to use of `x` in closure\n...\n33 |         })\n   |          - first borrow ends here\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
