{"message":"cannot borrow immutable item as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs","byte_start":862,"byte_end":869,"line_start":33,"line_end":33,"column_start":19,"column_end":26,"is_primary":true,"text":[{"text":"    let __isize = &mut *x; //~ ERROR cannot borrow","highlight_start":19,"highlight_end":26}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs:33:19\n   |\n33 |     let __isize = &mut *x; //~ ERROR cannot borrow\n   |                   ^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs","byte_start":950,"byte_end":957,"line_start":37,"line_end":37,"column_start":19,"column_end":26,"is_primary":true,"text":[{"text":"    let __isize = &mut *x; //~ ERROR cannot borrow","highlight_start":19,"highlight_end":26}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs:37:19\n   |\n37 |     let __isize = &mut *x; //~ ERROR cannot borrow\n   |                   ^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs","byte_start":1117,"byte_end":1125,"line_start":45,"line_end":45,"column_start":5,"column_end":13,"is_primary":true,"text":[{"text":"    &mut **x //~ ERROR cannot borrow","highlight_start":5,"highlight_end":13}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs:45:5\n   |\n45 |     &mut **x //~ ERROR cannot borrow\n   |     ^^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot borrow immutable item as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs","byte_start":1223,"byte_end":1231,"line_start":49,"line_end":49,"column_start":5,"column_end":13,"is_primary":true,"text":[{"text":"    &mut **x //~ ERROR cannot borrow","highlight_start":5,"highlight_end":13}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs:49:5\n   |\n49 |     &mut **x //~ ERROR cannot borrow\n   |     ^^^^^^^^ cannot borrow as mutable\n\n"}
{"message":"cannot assign to immutable item","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs","byte_start":1294,"byte_end":1300,"line_start":53,"line_end":53,"column_start":5,"column_end":11,"is_primary":true,"text":[{"text":"    *x = 3; //~ ERROR cannot assign","highlight_start":5,"highlight_end":11}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs:53:5\n   |\n53 |     *x = 3; //~ ERROR cannot assign\n   |     ^^^^^^ cannot mutate\n\n"}
{"message":"cannot assign to immutable item","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs","byte_start":1368,"byte_end":1375,"line_start":57,"line_end":57,"column_start":5,"column_end":12,"is_primary":true,"text":[{"text":"    **x = 3; //~ ERROR cannot assign","highlight_start":5,"highlight_end":12}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs:57:5\n   |\n57 |     **x = 3; //~ ERROR cannot assign\n   |     ^^^^^^^ cannot mutate\n\n"}
{"message":"cannot assign to immutable item","code":{"code":"E0594","explanation":null},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs","byte_start":1447,"byte_end":1454,"line_start":61,"line_end":61,"column_start":5,"column_end":12,"is_primary":true,"text":[{"text":"    **x = 3; //~ ERROR cannot assign","highlight_start":5,"highlight_end":12}],"label":"cannot mutate","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0594]: cannot assign to immutable item\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-borrow-overloaded-deref.rs:61:5\n   |\n61 |     **x = 3; //~ ERROR cannot assign\n   |     ^^^^^^^ cannot mutate\n\n"}
{"message":"aborting due to 7 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 7 previous errors\n\n"}
