{"message":"cannot borrow `p` as mutable because it is also borrowed as immutable (Ast)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":634,"byte_end":635,"line_start":22,"line_end":22,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    p[0] = 5; //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":623,"byte_end":624,"line_start":20,"line_end":20,"column_start":22,"column_end":23,"is_primary":false,"text":[{"text":"    let q: &isize = &p[0];","highlight_start":22,"highlight_end":23}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":693,"byte_end":694,"line_start":25,"line_end":25,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `p` as mutable because it is also borrowed as immutable (Ast)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":917,"byte_end":919,"line_start":37,"line_end":37,"column_start":9,"column_end":11,"is_primary":true,"text":[{"text":"        || p[0] = 5); //~ ERROR cannot borrow `p` as mutable","highlight_start":9,"highlight_end":11}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":906,"byte_end":907,"line_start":36,"line_end":36,"column_start":10,"column_end":11,"is_primary":false,"text":[{"text":"        &p,","highlight_start":10,"highlight_end":11}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":928,"byte_end":929,"line_start":37,"line_end":37,"column_start":20,"column_end":21,"is_primary":false,"text":[{"text":"        || p[0] = 5); //~ ERROR cannot borrow `p` as mutable","highlight_start":20,"highlight_end":21}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":920,"byte_end":921,"line_start":37,"line_end":37,"column_start":12,"column_end":13,"is_primary":false,"text":[{"text":"        || p[0] = 5); //~ ERROR cannot borrow `p` as mutable","highlight_start":12,"highlight_end":13}],"label":"borrow occurs due to use of `p` in closure","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `p` as mutable because it is also borrowed as immutable (Mir)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":634,"byte_end":635,"line_start":22,"line_end":22,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    p[0] = 5; //~ ERROR cannot borrow","highlight_start":5,"highlight_end":6}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":623,"byte_end":624,"line_start":20,"line_end":20,"column_start":22,"column_end":23,"is_primary":false,"text":[{"text":"    let q: &isize = &p[0];","highlight_start":22,"highlight_end":23}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":693,"byte_end":694,"line_start":25,"line_end":25,"column_start":1,"column_end":2,"is_primary":false,"text":[{"text":"}","highlight_start":1,"highlight_end":2}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `p` as mutable because it is also borrowed as immutable (Mir)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":917,"byte_end":919,"line_start":37,"line_end":37,"column_start":9,"column_end":11,"is_primary":true,"text":[{"text":"        || p[0] = 5); //~ ERROR cannot borrow `p` as mutable","highlight_start":9,"highlight_end":11}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":905,"byte_end":907,"line_start":36,"line_end":36,"column_start":9,"column_end":11,"is_primary":false,"text":[{"text":"        &p,","highlight_start":9,"highlight_end":11}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":928,"byte_end":929,"line_start":37,"line_end":37,"column_start":20,"column_end":21,"is_primary":false,"text":[{"text":"        || p[0] = 5); //~ ERROR cannot borrow `p` as mutable","highlight_start":20,"highlight_end":21}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-assign-comp-idx.rs","byte_start":920,"byte_end":921,"line_start":37,"line_end":37,"column_start":12,"column_end":13,"is_primary":false,"text":[{"text":"        || p[0] = 5); //~ ERROR cannot borrow `p` as mutable","highlight_start":12,"highlight_end":13}],"label":"borrow occurs due to use of `p` in closure","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
