{"message":"cannot borrow immutable item `x` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-addr-of-imm-var.rs","byte_start":525,"byte_end":531,"line_start":13,"line_end":13,"column_start":25,"column_end":31,"is_primary":true,"text":[{"text":"    let y: &mut isize = &mut x; //~ ERROR cannot borrow","highlight_start":25,"highlight_end":31}],"label":"cannot borrow as mutable","suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error[E0596]: cannot borrow immutable item `x` as mutable\n  --> /home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-mut-addr-of-imm-var.rs:13:25\n   |\n13 |     let y: &mut isize = &mut x; //~ ERROR cannot borrow\n   |                         ^^^^^^ cannot borrow as mutable\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
