{"message":"cannot borrow `v` as mutable because it is also borrowed as immutable (Ast)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-vec-content.rs","byte_start":959,"byte_end":961,"line_start":28,"line_end":28,"column_start":9,"column_end":11,"is_primary":true,"text":[{"text":"        || { //~ ERROR cannot borrow `v` as mutable","highlight_start":9,"highlight_end":11}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-vec-content.rs","byte_start":945,"byte_end":946,"line_start":27,"line_end":27,"column_start":10,"column_end":11,"is_primary":false,"text":[{"text":"        &v[0],","highlight_start":10,"highlight_end":11}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-vec-content.rs","byte_start":1034,"byte_end":1035,"line_start":30,"line_end":30,"column_start":10,"column_end":11,"is_primary":false,"text":[{"text":"        })","highlight_start":10,"highlight_end":11}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-vec-content.rs","byte_start":1015,"byte_end":1016,"line_start":29,"line_end":29,"column_start":13,"column_end":14,"is_primary":false,"text":[{"text":"            v[1] = 4;","highlight_start":13,"highlight_end":14}],"label":"borrow occurs due to use of `v` in closure","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"cannot borrow `v` as mutable because it is also borrowed as immutable (Mir)","code":{"code":"E0502","explanation":"\nThis error indicates that you are trying to borrow a variable as mutable when it\nhas already been borrowed as immutable.\n\nExample of erroneous code:\n\n```compile_fail,E0502\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    let ref y = a; // a is borrowed as immutable.\n    bar(a); // error: cannot borrow `*a` as mutable because `a` is also borrowed\n            //        as immutable\n}\n```\n\nTo fix this error, ensure that you don't have any other references to the\nvariable before trying to access it mutably:\n\n```\nfn bar(x: &mut i32) {}\nfn foo(a: &mut i32) {\n    bar(a);\n    let ref y = a; // ok!\n}\n```\n\nFor more information on the rust ownership system, take a look at\nhttps://doc.rust-lang.org/stable/book/references-and-borrowing.html.\n"},"level":"error","spans":[{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-vec-content.rs","byte_start":959,"byte_end":961,"line_start":28,"line_end":28,"column_start":9,"column_end":11,"is_primary":true,"text":[{"text":"        || { //~ ERROR cannot borrow `v` as mutable","highlight_start":9,"highlight_end":11}],"label":"mutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-vec-content.rs","byte_start":945,"byte_end":946,"line_start":27,"line_end":27,"column_start":10,"column_end":11,"is_primary":false,"text":[{"text":"        &v[0],","highlight_start":10,"highlight_end":11}],"label":"immutable borrow occurs here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-vec-content.rs","byte_start":1034,"byte_end":1035,"line_start":30,"line_end":30,"column_start":10,"column_end":11,"is_primary":false,"text":[{"text":"        })","highlight_start":10,"highlight_end":11}],"label":"immutable borrow ends here","suggested_replacement":null,"expansion":null},{"file_name":"/home/ariel/Rust/rust-master/src/test/compile-fail/borrowck/borrowck-loan-vec-content.rs","byte_start":1015,"byte_end":1016,"line_start":29,"line_end":29,"column_start":13,"column_end":14,"is_primary":false,"text":[{"text":"            v[1] = 4;","highlight_start":13,"highlight_end":14}],"label":"borrow occurs due to use of `v` in closure","suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}
